{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9eafea45-dfec-49c1-88ab-4f39233db3d2",
   "metadata": {},
   "source": [
    "# Hands on: Creating context-specific mechanistic networks from experimental data and prior knowledge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0310ca96-d68a-471e-bda7-d7c61312b3ba",
   "metadata": {},
   "source": [
    "This Python notebook serves as an example of how users can use their own datasets and integrate them with INDRA to provide meaningful interpretation.  We start with bringing our own MSstats dataset that consists of a list of proteins alongside their p-values.  We use p-values to filter which genes we query INDRA with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2fb8d3e7-9a3a-449c-af16-18323548d644",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install -q -r requirements.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3aad459-6315-4070-a1ba-bd5338deec81",
   "metadata": {},
   "source": [
    "## STEP 1: Import MSstats Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e775258-b1a7-4225-80a8-0ddb8fb48778",
   "metadata": {},
   "source": [
    "First, we will import MSstats datasets as pandas dataframes.  The MSstats dataset consist of the output of the MSstats groupComparison function, which consists of protein p-values.\n",
    "\n",
    "We filter the datasets to smaller sizes based on p-value.  You can adjust those parameters as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e4375a3d-4caf-4564-81de-efa09077851b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Label</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>SE</th>\n",
       "      <th>Tvalue</th>\n",
       "      <th>DF</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>adj.pvalue</th>\n",
       "      <th>issue</th>\n",
       "      <th>MissingPercentage</th>\n",
       "      <th>ImputationPercentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19890</th>\n",
       "      <td>BRD2_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>2.046185244</td>\n",
       "      <td>0.114339</td>\n",
       "      <td>17.895836</td>\n",
       "      <td>260.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.310067</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19935</th>\n",
       "      <td>BRD3_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>3.333427936</td>\n",
       "      <td>0.126571</td>\n",
       "      <td>26.336522</td>\n",
       "      <td>257.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.252349</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19980</th>\n",
       "      <td>BRD4_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>2.668934662</td>\n",
       "      <td>0.101283</td>\n",
       "      <td>26.351317</td>\n",
       "      <td>257.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.118121</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27900</th>\n",
       "      <td>CEBPZ_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.291058829</td>\n",
       "      <td>0.074340</td>\n",
       "      <td>-3.915236</td>\n",
       "      <td>260.0</td>\n",
       "      <td>1.154420e-04</td>\n",
       "      <td>3.005139e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.059060</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37530</th>\n",
       "      <td>CRNL1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.268053808</td>\n",
       "      <td>0.069816</td>\n",
       "      <td>-3.839419</td>\n",
       "      <td>260.0</td>\n",
       "      <td>1.549580e-04</td>\n",
       "      <td>3.832114e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027517</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41580</th>\n",
       "      <td>DAZP1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.617508071</td>\n",
       "      <td>0.099537</td>\n",
       "      <td>6.203818</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.159950e-09</td>\n",
       "      <td>1.780518e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.286577</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66060</th>\n",
       "      <td>FUBP2_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.291044501</td>\n",
       "      <td>0.077226</td>\n",
       "      <td>3.768747</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.030530e-04</td>\n",
       "      <td>4.782372e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017450</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66105</th>\n",
       "      <td>FUBP3_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.300409541</td>\n",
       "      <td>0.069882</td>\n",
       "      <td>4.298798</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.430235e-05</td>\n",
       "      <td>8.941156e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.095302</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72855</th>\n",
       "      <td>GTPB4_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.344028893</td>\n",
       "      <td>0.087806</td>\n",
       "      <td>-3.918077</td>\n",
       "      <td>260.0</td>\n",
       "      <td>1.141660e-04</td>\n",
       "      <td>3.005139e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.020134</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75735</th>\n",
       "      <td>HEAT3_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.409028514</td>\n",
       "      <td>0.100056</td>\n",
       "      <td>-4.087980</td>\n",
       "      <td>260.0</td>\n",
       "      <td>5.804139e-05</td>\n",
       "      <td>1.794204e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.127517</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77985</th>\n",
       "      <td>HNRPD_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.236133866</td>\n",
       "      <td>0.049974</td>\n",
       "      <td>4.725139</td>\n",
       "      <td>260.0</td>\n",
       "      <td>3.768760e-06</td>\n",
       "      <td>1.864029e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.142282</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92790</th>\n",
       "      <td>KRR1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.206019868</td>\n",
       "      <td>0.048706</td>\n",
       "      <td>-4.229846</td>\n",
       "      <td>260.0</td>\n",
       "      <td>3.242996e-05</td>\n",
       "      <td>1.069324e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010067</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119520</th>\n",
       "      <td>NOP2_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.260985294</td>\n",
       "      <td>0.045615</td>\n",
       "      <td>-5.721467</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.899095e-08</td>\n",
       "      <td>1.792365e-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.092617</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137745</th>\n",
       "      <td>PP1R8_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.358282468</td>\n",
       "      <td>0.069056</td>\n",
       "      <td>5.188314</td>\n",
       "      <td>260.0</td>\n",
       "      <td>4.275951e-07</td>\n",
       "      <td>2.349870e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140535</th>\n",
       "      <td>PRP18_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.485713404</td>\n",
       "      <td>0.120799</td>\n",
       "      <td>4.020834</td>\n",
       "      <td>260.0</td>\n",
       "      <td>7.603588e-05</td>\n",
       "      <td>2.212197e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.049664</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145440</th>\n",
       "      <td>QKI_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.336984673</td>\n",
       "      <td>0.055130</td>\n",
       "      <td>6.112567</td>\n",
       "      <td>260.0</td>\n",
       "      <td>3.570594e-09</td>\n",
       "      <td>2.522880e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196155</th>\n",
       "      <td>TIA1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.476791144</td>\n",
       "      <td>0.105361</td>\n",
       "      <td>4.525306</td>\n",
       "      <td>260.0</td>\n",
       "      <td>9.182867e-06</td>\n",
       "      <td>4.128951e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.058389</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210195</th>\n",
       "      <td>UTP15_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.59354411</td>\n",
       "      <td>0.079188</td>\n",
       "      <td>-7.495409</td>\n",
       "      <td>260.0</td>\n",
       "      <td>1.039169e-12</td>\n",
       "      <td>1.284932e-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214515</th>\n",
       "      <td>WDR12_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.217539679</td>\n",
       "      <td>0.049685</td>\n",
       "      <td>-4.378405</td>\n",
       "      <td>260.0</td>\n",
       "      <td>1.733909e-05</td>\n",
       "      <td>7.146595e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.001342</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214965</th>\n",
       "      <td>WDR43_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.502419318</td>\n",
       "      <td>0.070364</td>\n",
       "      <td>-7.140333</td>\n",
       "      <td>260.0</td>\n",
       "      <td>9.313883e-12</td>\n",
       "      <td>9.213293e-09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.144966</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215505</th>\n",
       "      <td>WDR75_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>-0.393941989</td>\n",
       "      <td>0.091846</td>\n",
       "      <td>-4.289155</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.530857e-05</td>\n",
       "      <td>8.941156e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.324161</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Protein       Label        log2FC        SE     Tvalue     DF  \\\n",
       "19890    BRD2_HUMAN  DMSO-DbET6   2.046185244  0.114339  17.895836  260.0   \n",
       "19935    BRD3_HUMAN  DMSO-DbET6   3.333427936  0.126571  26.336522  257.0   \n",
       "19980    BRD4_HUMAN  DMSO-DbET6   2.668934662  0.101283  26.351317  257.0   \n",
       "27900   CEBPZ_HUMAN  DMSO-DbET6  -0.291058829  0.074340  -3.915236  260.0   \n",
       "37530   CRNL1_HUMAN  DMSO-DbET6  -0.268053808  0.069816  -3.839419  260.0   \n",
       "41580   DAZP1_HUMAN  DMSO-DbET6   0.617508071  0.099537   6.203818  260.0   \n",
       "66060   FUBP2_HUMAN  DMSO-DbET6   0.291044501  0.077226   3.768747  260.0   \n",
       "66105   FUBP3_HUMAN  DMSO-DbET6   0.300409541  0.069882   4.298798  260.0   \n",
       "72855   GTPB4_HUMAN  DMSO-DbET6  -0.344028893  0.087806  -3.918077  260.0   \n",
       "75735   HEAT3_HUMAN  DMSO-DbET6  -0.409028514  0.100056  -4.087980  260.0   \n",
       "77985   HNRPD_HUMAN  DMSO-DbET6   0.236133866  0.049974   4.725139  260.0   \n",
       "92790    KRR1_HUMAN  DMSO-DbET6  -0.206019868  0.048706  -4.229846  260.0   \n",
       "119520   NOP2_HUMAN  DMSO-DbET6  -0.260985294  0.045615  -5.721467  260.0   \n",
       "137745  PP1R8_HUMAN  DMSO-DbET6   0.358282468  0.069056   5.188314  260.0   \n",
       "140535  PRP18_HUMAN  DMSO-DbET6   0.485713404  0.120799   4.020834  260.0   \n",
       "145440    QKI_HUMAN  DMSO-DbET6   0.336984673  0.055130   6.112567  260.0   \n",
       "196155   TIA1_HUMAN  DMSO-DbET6   0.476791144  0.105361   4.525306  260.0   \n",
       "210195  UTP15_HUMAN  DMSO-DbET6   -0.59354411  0.079188  -7.495409  260.0   \n",
       "214515  WDR12_HUMAN  DMSO-DbET6  -0.217539679  0.049685  -4.378405  260.0   \n",
       "214965  WDR43_HUMAN  DMSO-DbET6  -0.502419318  0.070364  -7.140333  260.0   \n",
       "215505  WDR75_HUMAN  DMSO-DbET6  -0.393941989  0.091846  -4.289155  260.0   \n",
       "\n",
       "              pvalue    adj.pvalue issue  MissingPercentage  \\\n",
       "19890   0.000000e+00  0.000000e+00   NaN           0.310067   \n",
       "19935   0.000000e+00  0.000000e+00   NaN           0.252349   \n",
       "19980   0.000000e+00  0.000000e+00   NaN           0.118121   \n",
       "27900   1.154420e-04  3.005139e-02   NaN           0.059060   \n",
       "37530   1.549580e-04  3.832114e-02   NaN           0.027517   \n",
       "41580   2.159950e-09  1.780518e-06   NaN           0.286577   \n",
       "66060   2.030530e-04  4.782372e-02   NaN           0.017450   \n",
       "66105   2.430235e-05  8.941156e-03   NaN           0.095302   \n",
       "72855   1.141660e-04  3.005139e-02   NaN           0.020134   \n",
       "75735   5.804139e-05  1.794204e-02   NaN           0.127517   \n",
       "77985   3.768760e-06  1.864029e-03   NaN           0.142282   \n",
       "92790   3.242996e-05  1.069324e-02   NaN           0.010067   \n",
       "119520  2.899095e-08  1.792365e-05   NaN           0.092617   \n",
       "137745  4.275951e-07  2.349870e-04   NaN           0.045638   \n",
       "140535  7.603588e-05  2.212197e-02   NaN           0.049664   \n",
       "145440  3.570594e-09  2.522880e-06   NaN           0.000000   \n",
       "196155  9.182867e-06  4.128951e-03   NaN           0.058389   \n",
       "210195  1.039169e-12  1.284932e-09   NaN           0.300000   \n",
       "214515  1.733909e-05  7.146595e-03   NaN           0.001342   \n",
       "214965  9.313883e-12  9.213293e-09   NaN           0.144966   \n",
       "215505  2.530857e-05  8.941156e-03   NaN           0.324161   \n",
       "\n",
       "        ImputationPercentage  \n",
       "19890                    0.0  \n",
       "19935                    0.0  \n",
       "19980                    0.0  \n",
       "27900                    0.0  \n",
       "37530                    0.0  \n",
       "41580                    0.0  \n",
       "66060                    0.0  \n",
       "66105                    0.0  \n",
       "72855                    0.0  \n",
       "75735                    0.0  \n",
       "77985                    0.0  \n",
       "92790                    0.0  \n",
       "119520                   0.0  \n",
       "137745                   0.0  \n",
       "140535                   0.0  \n",
       "145440                   0.0  \n",
       "196155                   0.0  \n",
       "210195                   0.0  \n",
       "214515                   0.0  \n",
       "214965                   0.0  \n",
       "215505                   0.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "P_VALUE_LOGFC_PATH = \"../data/model.csv\" # Set this path yourself\n",
    "LABELS_FILTER = [\"DMSO-DbET6\"]\n",
    "\n",
    "P_VALUE_FILTER = 0.05 # Adjust this yourself\n",
    "\n",
    "def construct_pvalue_logfc_df(filename):\n",
    "    \"\"\"Return a filtered data frame from the given data file.\"\"\"\n",
    "    pandas_df = pd.read_csv(filename)\n",
    "    pandas_df = pandas_df[pandas_df['issue'].isnull()]\n",
    "    pandas_df = pandas_df[pandas_df['adj.pvalue'] < P_VALUE_FILTER]\n",
    "    pandas_df = pandas_df[pandas_df['Label'].isin(LABELS_FILTER)]\n",
    "    return pandas_df\n",
    "\n",
    "pvalue_logfc_df = construct_pvalue_logfc_df(P_VALUE_LOGFC_PATH)\n",
    "pvalue_logfc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f63730f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DMSO-DbET6']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(set(pvalue_logfc_df.Label))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a5923d8-aa9b-4636-a53c-4c518205c28d",
   "metadata": {},
   "source": [
    "## STEP 2: ID CONVERSION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5abf4cc-3830-4066-becf-0bf1931e90ea",
   "metadata": {},
   "source": [
    "INDRA Cogex only accepts HGNC IDs at the time of this writing.  However, the dataset provided in the example above contains uniprot mnemonic IDs.  \n",
    "\n",
    "Luckily, INDRA has code to convert uniprot mnemonic IDs into HGNC ids. For now, we will store this mapping in a separate dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fbdeda02-bf1b-48b5-ae1c-8b19aec4c820",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1103'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from indra.databases import uniprot_client\n",
    "\n",
    "def uniprot_to_hgnc_id(uniprot_mnemonic):\n",
    "    \"\"\"Get an HGNC ID from a UniProt mnemonic.\"\"\"\n",
    "    uniprot_id = uniprot_client.get_id_from_mnemonic(uniprot_mnemonic)\n",
    "    if uniprot_id:\n",
    "        return uniprot_client.get_hgnc_id(uniprot_id)\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "uniprot_to_hgnc_id(\"BRD2_HUMAN\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e04aae1-648b-43a4-88be-721f07211b06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'BRD2'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def uniprot_to_hgnc_gene_name(uniprot_mnemonic):\n",
    "    \"\"\"Get an HGNC gene name from a UniProt mnemonic.\"\"\"\n",
    "    uniprot_id = uniprot_client.get_gene_name(uniprot_mnemonic)\n",
    "    return uniprot_id\n",
    "uniprot_to_hgnc_gene_name(\"BRD2_HUMAN\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b338f49d-fb69-43ac-995a-3fd4d2b4c5a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1103': 'BRD2_HUMAN',\n",
       " '1104': 'BRD3_HUMAN',\n",
       " '13575': 'BRD4_HUMAN',\n",
       " '24218': 'CEBPZ_HUMAN',\n",
       " '15762': 'CRNL1_HUMAN',\n",
       " '2683': 'DAZP1_HUMAN',\n",
       " '6316': 'FUBP2_HUMAN',\n",
       " '4005': 'FUBP3_HUMAN',\n",
       " '21535': 'GTPB4_HUMAN',\n",
       " '26087': 'HEAT3_HUMAN',\n",
       " '5036': 'HNRPD_HUMAN',\n",
       " '5176': 'KRR1_HUMAN',\n",
       " '7867': 'NOP2_HUMAN',\n",
       " '9296': 'PP1R8_HUMAN',\n",
       " '17351': 'PRP18_HUMAN',\n",
       " '21100': 'QKI_HUMAN',\n",
       " '11802': 'TIA1_HUMAN',\n",
       " '25758': 'UTP15_HUMAN',\n",
       " '14098': 'WDR12_HUMAN',\n",
       " '28945': 'WDR43_HUMAN',\n",
       " '25725': 'WDR75_HUMAN'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_hgnc_id_to_uniprot_mapping(pandas_df):\n",
    "    mappings = {}\n",
    "    for protein in pandas_df['Protein'].unique():\n",
    "        mappings[uniprot_to_hgnc_id(protein)] = protein\n",
    "    return mappings\n",
    "\n",
    "hgnc_id_to_uniprot_mapping = create_hgnc_id_to_uniprot_mapping(pvalue_logfc_df)\n",
    "hgnc_id_to_uniprot_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c5ef9b92-77fc-4dc5-8ec3-120089d85ee5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BRD2': 'BRD2_HUMAN',\n",
       " 'BRD3': 'BRD3_HUMAN',\n",
       " 'BRD4': 'BRD4_HUMAN',\n",
       " 'CEBPZ': 'CEBPZ_HUMAN',\n",
       " 'CRNKL1': 'CRNL1_HUMAN',\n",
       " 'DAZAP1': 'DAZP1_HUMAN',\n",
       " 'KHSRP': 'FUBP2_HUMAN',\n",
       " 'FUBP3': 'FUBP3_HUMAN',\n",
       " 'GTPBP4': 'GTPB4_HUMAN',\n",
       " 'HEATR3': 'HEAT3_HUMAN',\n",
       " 'HNRNPD': 'HNRPD_HUMAN',\n",
       " 'KRR1': 'KRR1_HUMAN',\n",
       " 'NOP2': 'NOP2_HUMAN',\n",
       " 'PPP1R8': 'PP1R8_HUMAN',\n",
       " 'PRPF18': 'PRP18_HUMAN',\n",
       " 'QKI': 'QKI_HUMAN',\n",
       " 'TIA1': 'TIA1_HUMAN',\n",
       " 'UTP15': 'UTP15_HUMAN',\n",
       " 'WDR12': 'WDR12_HUMAN',\n",
       " 'WDR43': 'WDR43_HUMAN',\n",
       " 'WDR75': 'WDR75_HUMAN'}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def create_hgnc_gene_name_to_uniprot_mapping(pandas_df):\n",
    "    mappings = {}\n",
    "    for protein in pandas_df['Protein'].unique():\n",
    "        mappings[uniprot_to_hgnc_gene_name(protein)] = protein\n",
    "    return mappings\n",
    "\n",
    "hgnc_gene_name_to_uniprot_mapping = create_hgnc_gene_name_to_uniprot_mapping(pvalue_logfc_df)\n",
    "hgnc_gene_name_to_uniprot_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e05ef7e2-ec8d-483b-be2f-366ebe49a1fb",
   "metadata": {},
   "source": [
    "## STEP 3: QUERY INDRA COGEX"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5756574f-e6e3-4e3a-a74a-0b7ec9a559a4",
   "metadata": {},
   "source": [
    "Using INDRA Cogex, we can extract subnetwork relationships among the proteins from the MSstats output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2b8e4124-e577-43af-b60b-912bfa266bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "def query_indra_subnetwork(groundings):\n",
    "    \"\"\"Return a list INDRA subnetwork relations based on a list of groundings.\"\"\"\n",
    "    res = requests.post(\n",
    "        'https://discovery.indra.bio/api/indra_subnetwork_relations',\n",
    "        json={'nodes': groundings}\n",
    "    )\n",
    "    return res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "35e368e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'data': {'belief': 0.65,\n",
       "  'evidence_count': 1,\n",
       "  'has_database_evidence': False,\n",
       "  'has_reader_evidence': True,\n",
       "  'has_retracted_evidence': False,\n",
       "  'medscan_only': False,\n",
       "  'source_counts': '{\"sparser\": 1}',\n",
       "  'sparser_only': True,\n",
       "  'stmt_hash': 6100415255007272,\n",
       "  'stmt_json': '{\"type\": \"Complex\", \"members\": [{\"name\": \"RAD21\", \"db_refs\": {\"UP\": \"O60216\", \"TEXT\": \"RAD21\", \"HGNC\": \"9811\", \"EGID\": \"5885\"}}, {\"name\": \"BRD2\", \"db_refs\": {\"UP\": \"P25440\", \"TEXT\": \"BRD2\", \"HGNC\": \"1103\", \"EGID\": \"6046\"}}, {\"name\": \"BRD4\", \"db_refs\": {\"UP\": \"O60885\", \"TEXT\": \"BRD4\", \"HGNC\": \"13575\", \"EGID\": \"23476\"}}], \"belief\": 0.65, \"evidence\": [{\"source_api\": \"sparser\", \"pmid\": \"28107481\", \"text\": \"However, we were unable to demonstrate any direct physical interaction between BRD2 or BRD4 with cohesin subunit RAD21 ( xref ), suggesting the conformational control of KSHV latency involves additional factors.\", \"annotations\": {\"found_by\": \"INTERACT\"}, \"text_refs\": {\"PMID\": \"28107481\", \"TRID\": 16352739, \"PMCID\": \"PMC5287475\", \"DOI\": \"10.1371/JOURNAL.PPAT.1006100\"}, \"source_hash\": -5857008756386512230}], \"id\": \"a2b3575a-ae54-4a6a-a256-726aefe5a387\", \"matches_hash\": \"6100415255007272\"}',\n",
       "  'stmt_type': 'Complex'},\n",
       " 'rel_type': 'indra_rel',\n",
       " 'source_id': '13575',\n",
       " 'source_ns': 'HGNC',\n",
       " 'target_id': '1103',\n",
       " 'target_ns': 'HGNC'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groundings = []\n",
    "for hgnc_id in hgnc_id_to_uniprot_mapping:\n",
    "    groundings.append(('HGNC', hgnc_id))\n",
    "subnetwork_relations = query_indra_subnetwork(groundings)\n",
    "subnetwork_relations[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4dd03aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = requests.post(\n",
    "        'https://discovery.indra.bio/api/get_stmts_for_stmt_hashes',\n",
    "        json={\n",
    "          \"stmt_hashes\": [\n",
    "            \"12484535149707124\",\n",
    "          ],\n",
    "          \"evidence_limit\": 200\n",
    "        }\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "92e8e9e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'belief': 0.99995,\n",
       "  'evidence': [{'annotations': {'agents': {'coords': [[110, 114], [119, 123]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '25603177',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -1268912701169983876,\n",
       "    'text': 'BRD2 and BRD4 have been reported to be involved in cell cycle progression, given the evidence that binding of BRD2 and BRD4 to acetylated chromatin persists even during mitosis when chromatin is highly condensed and transcription is interrupted [XREF_BIBR, XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS16011928',\n",
       "     'PMCID': 'PMC4307342',\n",
       "     'PMID': '25603177',\n",
       "     'TRID': 461769}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '25603177',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -431739944072706480,\n",
       "    'text': 'BRD2 and BRD4 have been reported to be involved in cell cycle progression, given the evidence that binding of BRD2 and BRD4 to acetylated chromatin persists even during mitosis when chromatin is highly condensed and transcription is interrupted [ xref , xref ].',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS16011928',\n",
       "     'PMCID': 'PMC4307342',\n",
       "     'PMID': '25603177',\n",
       "     'TRID': 461769}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28262505',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1362980574378353556,\n",
       "    'text': 'As an example, the differential binding of Brd2 and Brd4 in relationship to the enhancer features is illustrated with the ChIP-seq tracks for Il21 and Rock2 , two important genes in Th17 cell differentiation ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2016.12.022',\n",
       "     'MANUSCRIPT_ID': 'NIHMS843856',\n",
       "     'PII': 'S1097-2765(16)30865-6',\n",
       "     'PMCID': 'PMC5357147',\n",
       "     'PMID': '28262505',\n",
       "     'TRID': 17331828}},\n",
       "   {'annotations': {'agents': {'coords': [[45, 49], [54, 58]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000899'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28262505',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -3142412404303348316,\n",
       "    'text': 'These results suggested that the kinetics of Brd2 and Brd4 recruitment to their target gene enhancer sites is different, and Brd4 occupancy is likely temporally coupled to the transcriptional activation of these signature genes in Th17 cell differentiation.',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2016.12.022',\n",
       "     'MANUSCRIPT_ID': 'NIHMS843856',\n",
       "     'PII': 'S1097-2765(16)30865-6',\n",
       "     'PMCID': 'PMC5357147',\n",
       "     'PMID': '28262505',\n",
       "     'TRID': 17331828}},\n",
       "   {'annotations': {'agents': {'coords': [[151, 155], [156, 160]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_E031'},\n",
       "      'name': None},\n",
       "     'cell_type': {'db_refs': {'CL': 'CL:0000899'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28262505',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 2485932039899088822,\n",
       "    'text': 'Notably, the Brd2 and Brd4 co-bound genes exhibit highest expression level, while Brd2 only bound genes show modestly higher expression than genes without Brd2 and Brd4 binding (XREF_FIG).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2016.12.022',\n",
       "     'MANUSCRIPT_ID': 'NIHMS843856',\n",
       "     'PII': 'S1097-2765(16)30865-6',\n",
       "     'PMCID': 'PMC5357147',\n",
       "     'PMID': '28262505',\n",
       "     'TRID': 17331828}},\n",
       "   {'annotations': {'agents': {'coords': [[169, 173], [174, 178]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_3772'},\n",
       "      'name': None},\n",
       "     'cell_type': {'db_refs': {'CL': 'CL:0000899'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28262505',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 6833790319861018728,\n",
       "    'text': 'Similarly, peaks bound only by Brd2 are also enriched with Stat3, H3K27Ac and H3K4me1, correlating with the modest increase in gene expression compared to genes without Brd2 and Brd4 binding.',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2016.12.022',\n",
       "     'MANUSCRIPT_ID': 'NIHMS843856',\n",
       "     'PII': 'S1097-2765(16)30865-6',\n",
       "     'PMCID': 'PMC5357147',\n",
       "     'PMID': '28262505',\n",
       "     'TRID': 17331828}},\n",
       "   {'annotations': {'agents': {'coords': [[43, 47], [52, 56]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000899'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28262505',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 6687637843694227493,\n",
       "    'text': 'As an example, the differential binding of Brd2 and Brd4 in relationship to the enhancer features is illustrated with the ChIP-seq tracks for Il21 and Rock2, two important genes in Th17 cell differentiation (XREF_FIG).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2016.12.022',\n",
       "     'MANUSCRIPT_ID': 'NIHMS843856',\n",
       "     'PII': 'S1097-2765(16)30865-6',\n",
       "     'PMCID': 'PMC5357147',\n",
       "     'PMID': '28262505',\n",
       "     'TRID': 17331828}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '28666361',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -6196131680414509189,\n",
       "    'text': 'Efforts to directly assess the effect of Hat1 loss on the association of Brd2 and Brd4 with nascent chromatin have been hindered by technical difficulties arising from the effect of the iPOND procedure on the antigenicity of some proteins.',\n",
       "    'text_refs': {'DOI': '10.1093/NAR/GKX545',\n",
       "     'PMCID': 'PMC5766187',\n",
       "     'PMID': '28666361',\n",
       "     'TRID': 23428625}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '29448097',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 1273198415338768569,\n",
       "    'text': 'Degradation of BRD2 and BRD4 was associated with significantly reduced expression of several pro-inflammatory genes: inducible nitric oxide synthase (iNOS), cyclooxygenase-2 (COX-2), interleukin (IL)-1β, tumor necrosis factor-a (TNF-α), IL-6, chemokine (C-C motif) ligand 2 (CCL2), and matrix metalloproteinase-9 (MMP-9).',\n",
       "    'text_refs': {'DOI': '10.1016/J.BBRC.2018.02.096',\n",
       "     'PII': 'S0006-291X(18)30325-5',\n",
       "     'PMID': '29448097',\n",
       "     'TRID': 28663369}},\n",
       "   {'annotations': {'agents': {'coords': [[64, 68], [73, 77]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '24101376',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -7448167405094626904,\n",
       "    'text': 'Thus, BET bromodomain inhibition results in impaired binding of Brd2 and Brd4 to the Il17a gene and its suppressed transcription.',\n",
       "    'text_refs': {'DOI': '10.1084/JEM.20130376',\n",
       "     'PMCID': 'PMC3804955',\n",
       "     'PMID': '24101376',\n",
       "     'TRID': 14830058}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '24101376',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -5030966166575563806,\n",
       "    'text': 'After 20 h of culture, both Brd2 and Brd4 significantly associated with chromatin at the Il17a control region CNS2 (essential for Il17a transcription; xref ) in T H 17 but not T H 1 or T H 2 cells, and this binding was prevented by JQ1, significantly for Brd4 ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1084/JEM.20130376',\n",
       "     'PMCID': 'PMC3804955',\n",
       "     'PMID': '24101376',\n",
       "     'TRID': 14830058}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '24101376',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 3348722045115010747,\n",
       "    'text': 'JQ1 had no impact on H3K4me3 or acetylation on CNS2 ( xref ), suggesting that the reduced association of Brd2 and Brd4 at this site was not an indirect consequence of altered chromatin structure.',\n",
       "    'text_refs': {'DOI': '10.1084/JEM.20130376',\n",
       "     'PMCID': 'PMC3804955',\n",
       "     'PMID': '24101376',\n",
       "     'TRID': 14830058}},\n",
       "   {'annotations': {'agents': {'coords': [[148, 152], [157, 161]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_0006'},\n",
       "      'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'hypothesis': True, 'section_type': None},\n",
       "    'pmid': '24733848',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -1663058542856109591,\n",
       "    'text': 'By coimmunoprecipitation, BET proteins were found to be complexed with Nrf2, whereas chromatin immunoprecipitation studies indicated recruitment of Brd2 and Brd4 to Nrf2 binding sites on the promoters of heme oxygenase-1 and NADPH quinone oxidoreductase 1.',\n",
       "    'text_refs': {'DOI': '10.4049/JIMMUNOL.1301984',\n",
       "     'MANUSCRIPT_ID': 'EMS57649',\n",
       "     'PMCID': 'PMC4011694',\n",
       "     'PMID': '24733848',\n",
       "     'TRID': 15327612}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '24391744',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 7932191398242087428,\n",
       "    'text': 'RVX-208 binds to BRD4[BD1] and BRD2[BD2] quite differently than does JQ-1, which is based on a triazolodiazepine core scaffold ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PONE.0083190',\n",
       "     'PMCID': 'PMC3877016',\n",
       "     'PMID': '24391744',\n",
       "     'TRID': 12863164}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '24146617',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -3399788377626380698,\n",
       "    'text': 'The only prior evidence for a function of the residues within this patch was the observation that deletion of the last 23 amino acids of LANA abrogates interaction with BRD2 and BRD4.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1003672',\n",
       "     'PMCID': 'PMC3798644',\n",
       "     'PMID': '24146617',\n",
       "     'TRID': 18437976}},\n",
       "   {'annotations': {'agents': {'coords': [[32, 36], [55, 59]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '4932'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28710461',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 3325422374894809334,\n",
       "    'text': 'As expected, the recruitment of BRD4, BRD2 and BRD3 proteins was lost on nSBs in the presence of JQ1, thus validating the efficacy of the JQ1 treatment in stressed cells.',\n",
       "    'text_refs': {'DOI': '10.1038/S41598-017-05343-8',\n",
       "     'PMCID': 'PMC5511177',\n",
       "     'PMID': '28710461',\n",
       "     'TRID': 26399022}},\n",
       "   {'annotations': {'agents': {'coords': [[12, 16], [21, 25]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_F568'},\n",
       "      'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12305'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '20422051',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -7832113606310167420,\n",
       "    'text': 'Analysis of Brd2 and Brd4 recruitment during activation.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PONE.0010272',\n",
       "     'PMCID': 'PMC2858074',\n",
       "     'PMID': '20422051',\n",
       "     'TRID': 6018171}},\n",
       "   {'annotations': {'agents': {'coords': [[127, 131], [136, 140]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '12305'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '20422051',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 1297633045375091280,\n",
       "    'text': 'To investigate the timing of the events regulating transcriptional activation, we tracked the recruitment of GCN5, RNA pol II, Brd4 and Brd2 in relation to the VP16 activator, Cherry-tTA-ER.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PONE.0010272',\n",
       "     'PMCID': 'PMC2858074',\n",
       "     'PMID': '20422051',\n",
       "     'TRID': 6018171}},\n",
       "   {'annotations': {'agents': {'coords': [[35, 39], [47, 51]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '31324754',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 6832921576998854502,\n",
       "    'text': 'IFN-gamma increased the binding of BRD2, BRD3, BRD4, and Pol II on the IDO1 promoter by increasing the acetylation of histone H3, which could be prevented by JQ1 partially or even completely.',\n",
       "    'text_refs': {'DOI': '10.1038/S41419-019-1793-9',\n",
       "     'PMCID': 'PMC6642217',\n",
       "     'PMID': '31324754',\n",
       "     'TRID': 30445486}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '29603290',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -3681589234244299232,\n",
       "    'text': 'Brd2 and Brd4 associate with the IL17 locus in Th17 cells and regulate the expression of cytokines such as IL17, IL21 and GMCSF (Granulocyte-macrophage colony-stimulating factor) in experimental autoimmune encephalitis (EAE), a mouse model for MS[ xref ].',\n",
       "    'text_refs': {'DOI': '10.1002/BIES.201800007',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1510467',\n",
       "     'PMCID': 'PMC7025392',\n",
       "     'PMID': '29603290',\n",
       "     'TRID': 28436769}},\n",
       "   {'annotations': {'agents': {'coords': [[35, 39], [44, 48]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '32821285',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 2414767758529183005,\n",
       "    'text': 'This is likely a direct effect, as BRD2 and BRD4 recruitment to the CRP promoter is detected early on during dual cytokine stimulation (2h), and a short-term treatment with apabetalone or JQ1 (a pan-BETi with a different scaffold) counters the CRP promoter occupancy by BRD4 and the CRP gene induction by cytokines.',\n",
       "    'text_refs': {'DOI': '10.1155/2020/9397109',\n",
       "     'PMCID': 'PMC7416228',\n",
       "     'PMID': '32821285',\n",
       "     'TRID': 31998872}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '33105802',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 8130093037124285424,\n",
       "    'text': 'Brd2 and Brd4 physically associate with the promoters of inflammatory cytokine genes, reducing production of cytokines TNFα, IL-6, MCP-1 in vitro and in vivo.',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS21217838',\n",
       "     'PII': 'E7838',\n",
       "     'PMCID': 'PMC7660108',\n",
       "     'PMID': '33105802',\n",
       "     'TRID': 32300692}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '32723797',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 1824939100655059296,\n",
       "    'text': 'Furthermore, we showed that the E-protein interacts with bromodomain proteins (BRD4), which is in agreement with a previous study in which the researchers showed that SARS-CoV-2 envelop interacted with bromodomain proteins BRD2 and BRD4 to regulate gene transcription ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1128/MSYSTEMS.00505-20',\n",
       "     'PII': 'E00505-20',\n",
       "     'PMCID': 'PMC7394360',\n",
       "     'PMID': '32723797',\n",
       "     'TRID': 31894111}},\n",
       "   {'annotations': {'found_by': 'BIO-COMPLEX'},\n",
       "    'pmid': '30080437',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1370653684994509689,\n",
       "    'text': 'Fernandez-Alonso R, Davidson L, Hukelmann J, et al. Brd4-Brd2 isoform switching coordinates pluripotent exit and Smad2-dependent lineage specification.',\n",
       "    'text_refs': {'DOI': '10.1080/15592294.2018.1469891',\n",
       "     'PMCID': 'PMC6140815',\n",
       "     'PMID': '30080437',\n",
       "     'TRID': 29194343}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '31283444',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -4294641991660342779,\n",
       "    'text': 'The Bromodomain and Extra-Terminal Domain proteins Brd2 and Brd4 interact with the electrostatic patch on the LANA DBD ( xref , xref ).',\n",
       "    'text_refs': {'DOI': '10.1146/ANNUREV-VIROLOGY-092818-015716',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1591078',\n",
       "     'PMCID': 'PMC7415349',\n",
       "     'PMID': '31283444',\n",
       "     'TRID': 30403616}},\n",
       "   {'annotations': {'agents': {'coords': [[125, 129], [142, 146]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000235'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12721'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '24261714',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 4100179229960191050,\n",
       "    'text': 'It has recently been demonstrated that BRD2 is essential for proinflammatory cytokine production in macrophagesand that that BRD2, as well as BRD4, physically associates with promoters of inflammatory cytokine genes in macrophages [55].',\n",
       "    'text_refs': {'DOI': '10.1517/13543776.2014.859244',\n",
       "     'PMID': '24261714',\n",
       "     'TRID': 14409995}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 3231669455480182511,\n",
       "    'text': 'Additionally, the ET domains of BRD2 and BRD4 binding on the surface of KSHV LANA (kLANA) are required for the formation of kLANA “nuclear speckles” and latent replication [ xref ].',\n",
       "    'text_refs': {'DOI': '10.1186/S13567-020-00846-1',\n",
       "     'PMCID': 'PMC7509934',\n",
       "     'TRID': 32170661}},\n",
       "   {'annotations': {'agents': {'coords': [[189, 193], [198, 202]]},\n",
       "     'found_by': 'binding9'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10116'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '33664670',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 8972124502805479594,\n",
       "    'text': 'Since BET proteins can recognize acetylated lysine residues and interact with many chromatin remodeling proteins and some transcriptional factors, it is anticipated that the interaction of Brd2 and Brd4 with transcriptional factors and histone modifiers would initiate a signal that regulates expression of these two BET proteins through a feed-back loop, while inhibition of such interactions may block this process, leading to downregulation of Brd2 and Brd4.',\n",
       "    'text_refs': {'DOI': '10.3389/FPHAR.2021.636154',\n",
       "     'PMCID': 'PMC7921804',\n",
       "     'PMID': '33664670',\n",
       "     'TRID': 32889295}},\n",
       "   {'annotations': {'agents': {'coords': [[107, 111], [119, 123]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '33052224',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -8004389247056298804,\n",
       "    'text': 'Recently, the tumor promotion functions of bromodomain and extra-terminal domain (BET) proteins, including BRD2, BRD3, BRD4, and BRDT, which bind to acetylated histones, have attracted great attention XREF_BIBR.',\n",
       "    'text_refs': {'DOI': '10.7150/THNO.47432',\n",
       "     'PMCID': 'PMC7546000',\n",
       "     'PMID': '33052224',\n",
       "     'TRID': 32242050}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -4137554909019990108,\n",
       "    'text': 'The viral envelope E binds to BRD2 and BRD4 and could be the target for BET inhibitors [39, 45].',\n",
       "    'text_refs': {'DOI': '10.1186/S13148-020-00912-7',\n",
       "     'PMCID': 'PMC7404079',\n",
       "     'TRID': 31930067}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '32353859',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -558224296440506518,\n",
       "    'text': 'We found that the transmembrane E protein, likely resident on ERGIC and golgi membranes, binds to BRD2 and BRD4 ( xref ), members of the bromodomain and extra-terminal (BET) domain family of epigenetic readers that bind acetylated histones to regulate gene transcription  xref  .',\n",
       "    'text_refs': {'DOI': '10.1038/S41586-020-2286-9',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1587111',\n",
       "     'PMCID': 'PMC7431030',\n",
       "     'PMID': '32353859',\n",
       "     'TRID': 31491138}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -8051938457108164680,\n",
       "    'text': 'In this study we therefore explored (i) if the interaction of kLANA with Brd2 or Brd4 contributes to the previously reported preferential deposition of kLANA on cellular TSS and, (ii) if so, whether this would have an impact on kLANA-mediated transcriptional regulation.',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 7020387474289822725,\n",
       "    'text': 'Interestingly, treatment with I-BET151 markedly reduced the deposition of kLANA, as well as of Brd4 and Brd2, on cellular TSS, suggesting that the previously reported interaction of kLANA with Brd2 and Brd4 ( xref ; xref ; xref ; xref ) contributes to this characteristic association of kLANA with cellular TSS.',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 150519099135702065,\n",
       "    'text': 'To extend these observations to the related γ2-herpesvirus MHV-68 and to control for the well-known pleiotropic effects of I-BET151 on cellular gene expression ( xref ; xref ; xref ; xref ), we generated stable A20 murine B cell lines expressing either mLANA WT or mLANA mutant, in which three lysine residues in the basic patch on the surface of the mLANA DNA-binding domain had been mutated to alanine residues and which had lost the ability to interact with Brd2 and Brd4 ( xref ).',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 7912407618906750107,\n",
       "    'text': 'For mLANA-expressing murine B cells we found more cellular genes to be up- or downregulated compared to cells expressing the mLANA 3A mutant that is deficient in the interaction with Brd2 and Brd4 and which showed a reduced deposition of mLANA on cellular TSS ( xref , xref ); in addition, a significant proportion of mLANA-upregulated or -downregulated cellular genes had TSS that were occupied by mLANA ( xref ).',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -4706253506653637174,\n",
       "    'text': 'We used conventional ChIP-qPCR on selected cellular promoters of IQGAP3 and TRIM2 to confirm that the binding of kLANA as well as Brd2 and Brd4 was reduced in the presence of I-BET151 ( xref ).',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '33193257',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -2104842687209585569,\n",
       "    'text': 'We therefore transduced mouse A20 B cells with lentiviral vectors to express a FLAG-tagged mLANA WT or the previously described mLANA 3A mutant (K169A, K224A, and K228A), which is deficient in binding to Brd2 and Brd4 in vitro ( xref ), in a doxycycline-inducible manner ( xref ).',\n",
       "    'text_refs': {'DOI': '10.3389/FMICB.2020.591778',\n",
       "     'PMCID': 'PMC7642799',\n",
       "     'PMID': '33193257',\n",
       "     'TRID': 32395435}},\n",
       "   {'annotations': {'agents': {'coords': [[145, 149], [154, 158]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000056'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '33257646',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 4638557496233781811,\n",
       "    'text': 'BRD4 direct association to chromatin regulatory regions of the NADPH oxidase subunits increases in the mdx muscle and JQ1 administration reduces BRD4 and BRD2 recruitment at these regions.',\n",
       "    'text_refs': {'DOI': '10.1038/S41467-020-19839-X',\n",
       "     'PMCID': 'PMC7705749',\n",
       "     'PMID': '33257646',\n",
       "     'TRID': 32462298}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '33490899',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -8116242832074088240,\n",
       "    'text': 'Overall, these findings show that RANKL brings about changes in the BET protein chromatin-binding landscape at the very first stages of OC commitment (4h), chiefly detected as enhanced binding of Brd2 and Brd4, and such changes are exquisitely sensitive to pharmacological inhibition.',\n",
       "    'text_refs': {'DOI': '10.1016/J.ISCI.2020.101989',\n",
       "     'PMCID': 'PMC7807155',\n",
       "     'PMID': '33490899',\n",
       "     'TRID': 32703499}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '33490899',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 4689116051313994738,\n",
       "    'text': 'Gain of Brd2 or Brd4 binding occurs with, but mostly without, concurrent Brd4 and Brd2 increased binding, respectively, and in either case, it does not involve changes in Brd3 binding ( xref D–2F).',\n",
       "    'text_refs': {'DOI': '10.1016/J.ISCI.2020.101989',\n",
       "     'PMCID': 'PMC7807155',\n",
       "     'PMID': '33490899',\n",
       "     'TRID': 32703499}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33683199',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -3394885140870022069,\n",
       "    'text': 'However, recent results suggest that the acetyl-histone binding proteins BRD2 and BRD4 physically interact with the rate limiting origin firing factor TICRR/TRESLIN and that abrogation of this interaction disrupts the normal replication program (Sansam et al., 2018).',\n",
       "    'text_refs': {'DOI': '10.7554/ELIFE.62161',\n",
       "     'PII': 'E62161',\n",
       "     'PMCID': 'PMC7993996',\n",
       "     'PMID': '33683199',\n",
       "     'TRID': 32905265}},\n",
       "   {'annotations': {'found_by': 'BIO-FORM'},\n",
       "    'pmid': '32895492',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 3115544355397100819,\n",
       "    'text': 'These findings demonstrate that each BET protein forms nuclear condensates in definitive endoderm, and the condensates formed by BRD3 are distinct from those formed by BRD2 and BRD4.',\n",
       "    'text_refs': {'DOI': '10.1038/S41556-020-0572-2',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1672520',\n",
       "     'PMCID': 'PMC8008247',\n",
       "     'PMID': '32895492',\n",
       "     'TRID': 32075186}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '33717143',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -5337882370638047056,\n",
       "    'text': 'BET Bromodomain Inhibition Reduces Both BRD2 and BRD4 Binding to Promoters.',\n",
       "    'text_refs': {'DOI': '10.3389/FIMMU.2021.626255',\n",
       "     'PMCID': 'PMC7953504',\n",
       "     'PMID': '33717143',\n",
       "     'TRID': 32946310}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '32416067',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 2246333765927942016,\n",
       "    'text': 'We found that BRD4, BRD2, and H3K27ac bind to the same promoters and enhancers in both SUM159 and SUM149 cells ( xref and xref ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2020.04.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1589967',\n",
       "     'PII': 'S1097-2765(20)30269-0',\n",
       "     'PMCID': 'PMC7306005',\n",
       "     'PMID': '32416067',\n",
       "     'TRID': 31548524}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '32416067',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -3109462049206809725,\n",
       "    'text': 'Although both BRD4 and BRD2 binding was decreased after JQ1 treatment ( xref and xref ) and this was highly correlated in both promoter and enhancer regions ( xref and xref ), the decrease was more significant for BRD4 than BRD2.',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2020.04.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1589967',\n",
       "     'PII': 'S1097-2765(20)30269-0',\n",
       "     'PMCID': 'PMC7306005',\n",
       "     'PMID': '32416067',\n",
       "     'TRID': 31548524}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '32416067',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7372165254709077328,\n",
       "    'text': 'We observed an increase of BRD4 and BRD2 binding in the same promoter and enhancer regions in SUM159R cells compared to parental cells ( xref – xref and xref – xref ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2020.04.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1589967',\n",
       "     'PII': 'S1097-2765(20)30269-0',\n",
       "     'PMCID': 'PMC7306005',\n",
       "     'PMID': '32416067',\n",
       "     'TRID': 31548524}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '33355184',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1051051180124805933,\n",
       "    'text': 'Previous reports suggest this is through a physical interaction between GBAF-specific subunits and BRD4 [ xref , xref , xref – xref ], and in prostate cancers we find BRD9 interacts with BRD4 as well as BRD2.',\n",
       "    'text_refs': {'DOI': '10.1158/0008-5472.CAN-20-1417',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1656858',\n",
       "     'PII': 'CANRES.1417.2020',\n",
       "     'PMCID': 'PMC8026650',\n",
       "     'PMID': '33355184',\n",
       "     'TRID': 32572620}},\n",
       "   {'annotations': {'agents': {'coords': [[30, 34], [39, 43]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '34439792',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -8031645727601628275,\n",
       "    'text': 'Mutations in mLANA preventing BRD2 and BRD4 binding [158] also have similar consequences [98].',\n",
       "    'text_refs': {'DOI': '10.3390/BIOM11081126',\n",
       "     'PII': '1126',\n",
       "     'PMCID': 'PMC8391731',\n",
       "     'PMID': '34439792',\n",
       "     'TRID': 33700101}},\n",
       "   {'annotations': {'agents': {'coords': [[61, 65], [66, 70]]},\n",
       "     'found_by': 'binding11'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '31527644',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 6152522433921835495,\n",
       "    'text': 'JQ1 treatment did not affect the interactions between AR and BRD2/BRD4 (Fig.\\xa04A).',\n",
       "    'text_refs': {'DOI': '10.1038/S41598-019-49366-9',\n",
       "     'PMCID': 'PMC6746817',\n",
       "     'PMID': '31527644',\n",
       "     'TRID': 30647781}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '31504794',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -5286486223992926357,\n",
       "    'text': 'Although LYAR-UBF could associate with BRD2 or BRD4 even at basal cellular levels (Figure xref ), LYAR-SPT5 did not associate with BRD2 or BRD4 at all (Figure xref ).',\n",
       "    'text_refs': {'DOI': '10.1093/NAR/GKZ747',\n",
       "     'PII': 'GKZ747',\n",
       "     'PMCID': 'PMC6821171',\n",
       "     'PMID': '31504794',\n",
       "     'TRID': 30622442}},\n",
       "   {'annotations': {'found_by': 'BIO-COMPLEX'},\n",
       "    'pmid': '31504794',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7062373494007330051,\n",
       "    'text': 'Collectively, these data support the previous notice that LYAR forms a complex with BRD2 or BRD4, after which either complex can be recruited to rDNA via UBF.',\n",
       "    'text_refs': {'DOI': '10.1093/NAR/GKZ747',\n",
       "     'PII': 'GKZ747',\n",
       "     'PMCID': 'PMC6821171',\n",
       "     'PMID': '31504794',\n",
       "     'TRID': 30622442}},\n",
       "   {'annotations': {'found_by': 'BIO-COMPLEX'},\n",
       "    'pmid': '34199719',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 1005713163275066452,\n",
       "    'text': 'In Th17 immune cells, BRD2 interacts with the chromatin insulator CTCF to facilitate SE formation, BRD4 recruitment and gene expression changes via direct BRD2-BRD4 interaction [ xref ].',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS22116059',\n",
       "     'PII': '6059',\n",
       "     'PMCID': 'PMC8199980',\n",
       "     'PMID': '34199719',\n",
       "     'TRID': 33455506}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '32151278',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1649528349232352660,\n",
       "    'text': 'Interestingly, the BRG1 component of the SWI/SNF complex which is present in the MITF interactome is also known to physically interact with BRD4 and BRD2 [ xref , xref ].',\n",
       "    'text_refs': {'DOI': '10.1186/S13072-020-00333-Z',\n",
       "     'PMCID': 'PMC7063807',\n",
       "     'PMID': '32151278',\n",
       "     'TRID': 31265403}},\n",
       "   {'annotations': {'found_by': 'BIO-COMPLEX'},\n",
       "    'pmid': '36188615',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -2015585232763465486,\n",
       "    'text': 'In recent years, BET family-targeting small molecules (BRD2–BRD4 and testis-specific BRDT) are novel epi-drugs.',\n",
       "    'text_refs': {'DOI': '10.3389/FPHAR.2022.946811',\n",
       "     'PMCID': 'PMC9522521',\n",
       "     'PMID': '36188615',\n",
       "     'TRID': 35557210}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '36187481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -6177395182117242136,\n",
       "    'text': 'In 2013, Picaud et al. found that PFI-1 inhibitors were acetyl lysine mimics and could replace acetyl lysine to bind to BRD2 and BRD4, thereby inhibiting the recruitment of BET protein to chromatin, down-regulating the expression of MYC and Aurora B kinases ( xref ).',\n",
       "    'text_refs': {'DOI': '10.3389/FCELL.2022.1021820',\n",
       "     'PMCID': 'PMC9523081',\n",
       "     'PMID': '36187481',\n",
       "     'TRID': 35553948}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '32946883',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 1488006042622618533,\n",
       "    'text': 'In the 6-OHDA rodent model of levodopa-induced dyskinesia, an animal model for Parkinson’s disease, increased levels of Brd2 mRNA were observed in the dorsal striatum, along with increased binding of BRD2 and BRD4 proteins to the promoter and enhancer regions of the genes dysregulated during dyskinesia induction (e.g., Dab1 , Esr1 , Ntrk2 , Arc , FosB , and Nedd41 ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.NEUROPHARM.2020.108306',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1633343',\n",
       "     'PII': 'S0028-3908(20)30374-9',\n",
       "     'PMCID': 'PMC7655633',\n",
       "     'PMID': '32946883',\n",
       "     'TRID': 32128409}},\n",
       "   {'annotations': {'agents': {'coords': [[14, 18], [23, 27]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_5423'},\n",
       "      'name': None},\n",
       "     'cell_type': {'db_refs': {'CL': 'CL:0000988'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '32416067',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 1015099416024236278,\n",
       "    'text': 'Although both BRD4 and BRD2 binding was decreased after JQ1 treatment (Figures 5B and S5B) and this was highly correlated in both promoter and enhancer regions (Figures 5C and S5C), the decrease was more significant for BRD4 than BRD2.',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2020.04.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1589967',\n",
       "     'PII': 'S1097-2765(20)30269-0',\n",
       "     'PMCID': 'PMC7306005',\n",
       "     'PMID': '32416067',\n",
       "     'TRID': 31548524}},\n",
       "   {'annotations': {'agents': {'coords': [[27, 31], [36, 40]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_5423'},\n",
       "      'name': None},\n",
       "     'cell_type': {'db_refs': {'CL': 'CL:0000988'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '32416067',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 5870830065324646975,\n",
       "    'text': 'We observed an increase of BRD4 and BRD2 binding in the same promoter and enhancer regions in SUM159R cells compared to parental cells (Figures 5B–C and S7B–C).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2020.04.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1589967',\n",
       "     'PII': 'S1097-2765(20)30269-0',\n",
       "     'PMCID': 'PMC7306005',\n",
       "     'PMID': '32416067',\n",
       "     'TRID': 31548524}},\n",
       "   {'annotations': {'agents': {'coords': [[69, 73], [84, 88]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '31504794',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 5195737208917340567,\n",
       "    'text': 'Collectively, these data suggested that LYAR-mediated recruitment of BRD2, KAT7 and BRD4 to rDNA controls the acetylation of local histones H3 and H4 to cause the relaxation of these histones from rDNA loci and thereby promote rDNA transcription.',\n",
       "    'text_refs': {'DOI': '10.1093/NAR/GKZ747',\n",
       "     'PII': 'GKZ747',\n",
       "     'PMCID': 'PMC6821171',\n",
       "     'PMID': '31504794',\n",
       "     'TRID': 30622442}},\n",
       "   {'annotations': {'agents': {'coords': [[127, 131], [136, 140]]},\n",
       "     'found_by': 'binding1forms'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '31504794',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 5055445929519836335,\n",
       "    'text': 'Conversely, a pulldown assay using BRD2 or BRD4 as bait showed that neither protein could associate with SPT5 or UBF; however, BRD2 and BRD4—independently of each other—could form a complex containing both LYAR and KAT7-JADE3 (Figure 7C).',\n",
       "    'text_refs': {'DOI': '10.1093/NAR/GKZ747',\n",
       "     'PII': 'GKZ747',\n",
       "     'PMCID': 'PMC6821171',\n",
       "     'PMID': '31504794',\n",
       "     'TRID': 30622442}},\n",
       "   {'annotations': {'agents': {'coords': [[76, 80], [85, 89]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '32151278',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -3889510220238043202,\n",
       "    'text': 'Therefore, BRG1 or a BRG1-associated factor (BAF) could potentially mediate BRD4 and BRD2 interactions with MITF.',\n",
       "    'text_refs': {'DOI': '10.1186/S13072-020-00333-Z',\n",
       "     'PMCID': 'PMC7063807',\n",
       "     'PMID': '32151278',\n",
       "     'TRID': 31265403}},\n",
       "   {'annotations': {'agents': {'coords': [[69, 73], [78, 82]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '34614283',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -2319773555182221260,\n",
       "    'text': 'We found this “linker\\neffect” even more pronounced in the context of BRD2 and BRD4 binding.',\n",
       "    'text_refs': {'DOI': '10.1002/ANIE.202108848',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1746931',\n",
       "     'PMCID': 'PMC8648984',\n",
       "     'PMID': '34614283',\n",
       "     'TRID': 33903018}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '30845782',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -8250698883435670045,\n",
       "    'text': 'A short motif in the large C-terminus mediates the binding of Treslin/TICRR Sld3 to the BRD2 and BRD4 proteins that associate with acetylated chromatin via bromodomains.',\n",
       "    'text_refs': {'DOI': '10.3390/GENES10030199',\n",
       "     'PII': 'E199',\n",
       "     'PMCID': 'PMC6470937',\n",
       "     'PMID': '30845782',\n",
       "     'TRID': 30006704}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '36252792',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7129274209896767760,\n",
       "    'text': 'The SARS-CoV-2 E protein has been shown to interact with BRD4 and BRD2.',\n",
       "    'text_refs': {'DOI': '10.1016/J.CELLSIG.2022.110495',\n",
       "     'PII': 'S0898-6568(22)00257-1',\n",
       "     'PMCID': 'PMC9568271',\n",
       "     'PMID': '36252792',\n",
       "     'TRID': 35622609}},\n",
       "   {'annotations': {'agents': {'coords': [[77, 81], [83, 87]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0002322'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '35410381',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -7615547179400382531,\n",
       "    'text': 'Finally, we found that BRD3 depletion did not appear to significantly impact BRD2, BRD4 or CTCF binding to chromatin (Extended Data Fig.10b), possibly due to the fact that BRD3 is expressed at significantly lower levels than BRD2 and BRD4 in ESCs (Fig.4i).',\n",
       "    'text_refs': {'DOI': '10.1038/S41588-022-01044-9',\n",
       "     'MANUSCRIPT_ID': 'HHMIMS1798525',\n",
       "     'PMCID': 'PMC9099420',\n",
       "     'PMID': '35410381',\n",
       "     'TRID': 34783711}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '34614283',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1574796997179142856,\n",
       "    'text': 'A similar increase in the CRBN affinity of dBET1 compared\\nto thalidomide has been recently reported by others.[ xref ]We found this “linker\\neffect” even more pronounced in the context of BRD2 and BRD4 binding.',\n",
       "    'text_refs': {'DOI': '10.1002/ANIE.202108848',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1746931',\n",
       "     'PMCID': 'PMC8648984',\n",
       "     'PMID': '34614283',\n",
       "     'TRID': 33903018}},\n",
       "   {'annotations': {'agents': {'coords': [[0, 4], [20, 24]]},\n",
       "     'found_by': 'binding1forms'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '36793283',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -9198286567794486302,\n",
       "    'text': 'BRD2 ,  BRD3 , and  BRD4  and their neighboring genes form a complex network of interactions mainly through co-expression, physical interactions, and shared protein domains.',\n",
       "    'text_refs': {'DOI': '10.3389/FENDO.2023.1089531',\n",
       "     'PMCID': 'PMC9922706',\n",
       "     'PMID': '36793283',\n",
       "     'TRID': 36236955}},\n",
       "   {'annotations': {'agents': {'coords': [[187, 191], [196, 200]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000785'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '26654227',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -8794736254219396140,\n",
       "    'text': 'I-BET151, another novel highly selective inhibitor of both BRD2 and BRD4 causes transcriptional silencing of the c-MYC gene and c-MYC-dependent programs in myeloma cells by inhibition of BRD2 and BRD4 binding to the promoter of the c-MYC gene and failure to recruit P-TEFb and PAFc1 complexes to the IGH and c-MYC enhancer by BRD2 and BRD4, and lack of subsequent RNA-Pol-II recruitment [XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.1186/S12943-015-0474-2',\n",
       "     'PMCID': 'PMC4676894',\n",
       "     'PMID': '26654227',\n",
       "     'TRID': 18648644}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '24954007',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -4366907290921538505,\n",
       "    'text': 'Notably, Olinone showed modest binding to the bromodomains from CBP and PHIP withK d of 33 μM and 103.4 μM, respectively, very weak binding to BRD4-BrD2 withK d >300 μM, and almost no detectable binding to the bromodomains from many other proteins including BRD7, PCAF, ASH1L, TAF1L, SMARCA4, BAZ1B, BAZ2B, ATAD2, BPTF and TRIM24 that represent different subgroups of the human bromodomain family ( xref and xref ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.CHEMBIOL.2014.05.009',\n",
       "     'MANUSCRIPT_ID': 'NIHMS602357',\n",
       "     'PMCID': 'PMC4104156',\n",
       "     'PMID': '24954007',\n",
       "     'TRID': 4631772}},\n",
       "   {'annotations': {'agents': {'coords': [[22, 26], [31, 35]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '25606026',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 952004556399021338,\n",
       "    'text': 'BRD proteins, such as Brd2 and Brd4, which bind acetylated residues on histones and transcription factors, are important regulators of transcriptional elongation of a subset of inflammatory genes with paused polymerase II [XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.1186/S12950-014-0043-2',\n",
       "     'PMCID': 'PMC4299484',\n",
       "     'PMID': '25606026',\n",
       "     'TRID': 461366}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28949335',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 3823531228494025132,\n",
       "    'text': 'The Bromodomain Extraterminal (BET) subfamily of chromatin readers is highly specific toward acetylated histone tails, with highest affinity for H4K12, and H4K5. xref , xref This is interesting, as BRD2 and BRD4 preferentially bind H4K12ac. xref , xref Moreover, H4K5ac and H4K12ac were specifically linked to memory function and age-associated memory impairment. xref , xref , xref ',\n",
       "    'text_refs': {'DOI': '10.1038/TP.2017.202',\n",
       "     'PMCID': 'PMC5639246',\n",
       "     'PMID': '28949335',\n",
       "     'TRID': 7437585}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '18406326',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 6834652193375750423,\n",
       "    'text': 'The Brd2 and Brd4 proteins associate primarily with euchromatic regions of the genome, and are largely excluded from heterochromatin, suggesting that these proteins might regulate transcription ( Crowley et\\xa0al., 2002; Dey et\\xa0al., 2000, 2003; Mattsson et\\xa0al., 2002 ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2008.01.018',\n",
       "     'PMCID': 'PMC2387119',\n",
       "     'PMID': '18406326',\n",
       "     'TRID': 27222031}},\n",
       "   {'annotations': {'author': 'Rahman S (2011)',\n",
       "     'biogrid_a': '117036',\n",
       "     'biogrid_b': '111973',\n",
       "     'biogrid_int_id': '612035',\n",
       "     'entrez_a': '23476',\n",
       "     'entrez_b': '6046',\n",
       "     'exp_system': 'Affinity Capture-MS',\n",
       "     'exp_system_type': 'physical',\n",
       "     'hgnc_a': 'BRD4',\n",
       "     'hgnc_b': 'BRD2',\n",
       "     'modification': None,\n",
       "     'organism_a': '9606',\n",
       "     'organism_b': '9606',\n",
       "     'phenotypes': None,\n",
       "     'pmid': '21555454',\n",
       "     'qualifications': None,\n",
       "     'score': None,\n",
       "     'source_db': 'BIOGRID',\n",
       "     'syn_a': 'CAP|HUNK1|HUNKI|MCAP',\n",
       "     'syn_b': 'D6S113E|FSH|FSRG1|NAT|RING3|RNF3',\n",
       "     'syst_name_a': None,\n",
       "     'syst_name_b': 'DADB-17J1.3',\n",
       "     'tags': None,\n",
       "     'throughput': 'High Throughput'},\n",
       "    'pmid': '21555454',\n",
       "    'source_api': 'biogrid',\n",
       "    'source_hash': 4060451246017069409,\n",
       "    'source_id': '612035',\n",
       "    'text_refs': {'PMID': '21555454'}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '31126889',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7452390478202661618,\n",
       "    'text': 'Mechanistically, ChIP assays demonstrated that JQ1 decreased the association of BRD4 and BRD2 with promoter loci of Ku80 and RAD51, and shRNA data showed that expression of Ku80 and RAD51 was BRD4- and BRD2-dependent in PDAC cell lines.',\n",
       "    'text_refs': {'DOI': '10.1016/J.EBIOM.2019.05.035',\n",
       "     'PII': 'S2352-3964(19)30343-3',\n",
       "     'PMCID': 'PMC6604668',\n",
       "     'PMID': '31126889',\n",
       "     'TRID': 30243534}},\n",
       "   {'annotations': {'agents': {'coords': [[122, 126], [138, 142]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28388437',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 2650534930676615631,\n",
       "    'text': 'Consistent with previous reports, we also noted some places, such as the alpha-globin genes (Hba), at which BRD2, BRD3, and BRD4 were all bound (XREF_FIG).',\n",
       "    'text_refs': {'DOI': '10.1016/J.MOLCEL.2017.02.027',\n",
       "     'MANUSCRIPT_ID': 'NIHMS858015',\n",
       "     'PII': 'S1097-2765(17)30164-8',\n",
       "     'PMCID': 'PMC5393350',\n",
       "     'PMID': '28388437',\n",
       "     'TRID': 8522688}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '19736624',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -4345452178509464260,\n",
       "    'text': 'Additionally, both BRD4 and BRD2 interact with the murine γ-herpesvirus 68 protein or f73, which is required to establish viral latency in vivo [ xref ].',\n",
       "    'text_refs': {'MANUSCRIPT_ID': 'NIHMS224823',\n",
       "     'PMCID': 'PMC2921942',\n",
       "     'PMID': '19736624',\n",
       "     'TRID': 11175778}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28176838',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1087395168170410616,\n",
       "    'text': 'DMA inhibits the binding activity of BRD2 and BRD4 ( xref ) with an IC50 value of 11\\u2009mM and 6\\u2009mM, respectively ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1038/SREP42108',\n",
       "     'PMCID': 'PMC5296751',\n",
       "     'PMID': '28176838',\n",
       "     'TRID': 2658089}},\n",
       "   {'annotations': {'agents': {'coords': [[82, 86], [97, 101]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_0042'},\n",
       "      'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'hypothesis': True, 'section_type': None},\n",
       "    'pmid': '23826629',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 5234946537651563510,\n",
       "    'text': 'Enrichment for specific acetylated patterns may have interesting implications for Brd2, Brd3 and Brd4 recruitment, as discussed later.',\n",
       "    'text_refs': {'DOI': '10.1186/1756-8935-6-20',\n",
       "     'PMCID': 'PMC3710262',\n",
       "     'PMID': '23826629',\n",
       "     'TRID': 17701108}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 2577232384374050208,\n",
       "    'text': 'BRD2 and BRD4 binds regulatory regions of KSHV latent episome.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '30266203',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7088222734208743476,\n",
       "    'text': 'One of the critical origin firing factors – treslin/TICRR – has been shown to directly interact with the acetyl-histone binding proteins BRD2 and BRD4.',\n",
       "    'text_refs': {'DOI': '10.1016/J.DNAREP.2018.09.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1508100',\n",
       "     'PII': 'S1568-7864(18)30219-2',\n",
       "     'PMCID': 'PMC6261693',\n",
       "     'PMID': '30266203',\n",
       "     'TRID': 29385036}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '31553917',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1654459164611243106,\n",
       "    'text': 'In agreement with our data, protein interaction screens previously suggested interactions between BRD2 and BRD4 with replication factor C subunits ( xref ; xref ; xref ).',\n",
       "    'text_refs': {'DOI': '10.1016/J.CELREP.2019.08.051',\n",
       "     'MANUSCRIPT_ID': 'NIHMS1543783',\n",
       "     'PII': 'S2211-1247(19)31099-X',\n",
       "     'PMCID': 'PMC6878991',\n",
       "     'PMID': '31553917',\n",
       "     'TRID': 30671664}},\n",
       "   {'annotations': {'agents': {'coords': [[11, 15], [16, 20]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'location': {'db_refs': {'GO': 'GO:0005739'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '27008626',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 5801615905521357148,\n",
       "    'text': 'Binding of Brd2 and Brd4 may prevent the recruitment of a transcription factor (s) that activates insulin transcription [XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PONE.0151329',\n",
       "     'PMCID': 'PMC4805167',\n",
       "     'PMID': '27008626',\n",
       "     'TRID': 19647162}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '27110299',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 2160308729922722022,\n",
       "    'text': 'NMP inhibited BRD2 and BRD4 binding in an AlphaScreen assay, with an IC50 of 3 and 4\\xa0mM, respectively.',\n",
       "    'text_refs': {'DOI': '10.1186/S13148-016-0209-2',\n",
       "     'PMCID': 'PMC4840488',\n",
       "     'PMID': '27110299',\n",
       "     'TRID': 22432025}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '27110299',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -383841756040782507,\n",
       "    'text': 'To assess the possibility of NMP acting as a bromodomain inhibitor particularly for BRD2 and BRD4, both associated with adipogenesis, we used an AlphaScreen assay format to determine the effect of NMP on recombinant human BET bromodomains.',\n",
       "    'text_refs': {'DOI': '10.1186/S13148-016-0209-2',\n",
       "     'PMCID': 'PMC4840488',\n",
       "     'PMID': '27110299',\n",
       "     'TRID': 22432025}},\n",
       "   {'annotations': {'agents': {'coords': [[14, 18], [23, 27]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '27110299',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 7021913188222329705,\n",
       "    'text': 'NMP inhibited BRD2 and BRD4 binding in an AlphaScreen assay, with an IC50 of 3 and 4mM, respectively.',\n",
       "    'text_refs': {'DOI': '10.1186/S13148-016-0209-2',\n",
       "     'PMCID': 'PMC4840488',\n",
       "     'PMID': '27110299',\n",
       "     'TRID': 22432025}},\n",
       "   {'annotations': {'agents': {'coords': [[90, 94], [99, 103]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28286180',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 1562642621451072895,\n",
       "    'text': 'We found that he promoter regions of Dab1, Esr1, and Ntrk2 exhibited enhancements in both Brd2 and Brd4 binding in an L-DOPA dependent manner (XREF_FIG, Unpaired t-test, Dab1 : p = 0.007; Esr1 : p = 0.004; Ntrk2 : p = 0.034) (XREF_FIG, Unpaired t-test, Dab1 : p = 0.012; Esr1 : p = 0.004; Ntrk2 : p = 0.006).',\n",
       "    'text_refs': {'DOI': '10.1016/J.NBD.2017.03.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS861546',\n",
       "     'PII': 'S0969-9961(17)30050-5',\n",
       "     'PMCID': 'PMC5410664',\n",
       "     'PMID': '28286180',\n",
       "     'TRID': 17321672}},\n",
       "   {'annotations': {'agents': {'coords': [[135, 139], [144, 148]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28286180',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 9084938997751471487,\n",
       "    'text': \"To evaluate JQ1 's effect on BET protein function, we then used ChIP to determine if our previously identified LID dependent changes in Brd2 and Brd4 DNA binding were inhibited by the JQ1 treatment paradigm.\",\n",
       "    'text_refs': {'DOI': '10.1016/J.NBD.2017.03.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS861546',\n",
       "     'PII': 'S0969-9961(17)30050-5',\n",
       "     'PMCID': 'PMC5410664',\n",
       "     'PMID': '28286180',\n",
       "     'TRID': 17321672}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28286180',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -6648248968870450615,\n",
       "    'text': 'We found that he promoter regions of Dab1, Esr1, and Ntrk2 exhibited enhancements in both Brd2 and Brd4 binding in an L-DOPA dependent manner ( xref , Unpaired t-test, Dab1: p = 0.007; Esr1: p = 0.004; Ntrk2: p = 0.034) ( xref , Unpaired t-test, Dab1: p = 0.012; Esr1: p = 0.004; Ntrk2: p = 0.006).',\n",
       "    'text_refs': {'DOI': '10.1016/J.NBD.2017.03.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS861546',\n",
       "     'PII': 'S0969-9961(17)30050-5',\n",
       "     'PMCID': 'PMC5410664',\n",
       "     'PMID': '28286180',\n",
       "     'TRID': 17321672}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '28286180',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 2226218782658567623,\n",
       "    'text': 'Similarly, predicted enhancer regions of Arc , FosB, and Nedd4l also showed increases in the association of Brd2 and Brd4 following repeated L-DOPA administration ( xref , Unpaired t-test, Arc: p = 0.038; FosB: p = 0.002; Nedd4l: p = 0.002) ( xref , Arc: p = 0.022; FosB: p = 0.001; Nedd4l: p = 0.008).',\n",
       "    'text_refs': {'DOI': '10.1016/J.NBD.2017.03.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS861546',\n",
       "     'PII': 'S0969-9961(17)30050-5',\n",
       "     'PMCID': 'PMC5410664',\n",
       "     'PMID': '28286180',\n",
       "     'TRID': 17321672}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28286180',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -3890209657830921265,\n",
       "    'text': 'We observed an enhancement in the transcriptional expression of Brd2 following the development of dyskinetic behaviors that correlated with increases in locus-specific Brd2 and Brd4 binding in the dorsal striatum.',\n",
       "    'text_refs': {'DOI': '10.1016/J.NBD.2017.03.003',\n",
       "     'MANUSCRIPT_ID': 'NIHMS861546',\n",
       "     'PII': 'S0969-9961(17)30050-5',\n",
       "     'PMCID': 'PMC5410664',\n",
       "     'PMID': '28286180',\n",
       "     'TRID': 17321672}},\n",
       "   {'annotations': {'agents': {'coords': [[118, 122], [127, 131]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0009002'}, 'name': None},\n",
       "     'location': {'db_refs': {'GO': 'GO:0018995'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '29167426',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -6206896914022644068,\n",
       "    'text': 'We revealed a direct role for BET proteins both in skeletal muscle and in the tumor, by showing direct recruitment of BRD4 and BRD2 at a subset of pro atrophic genes and pro cachectic mediators.',\n",
       "    'text_refs': {'DOI': '10.1038/S41467-017-01645-7',\n",
       "     'PMCID': 'PMC5700099',\n",
       "     'PMID': '29167426',\n",
       "     'TRID': 3293711}},\n",
       "   {'annotations': {'agents': {'coords': [[216, 220], [225, 229]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_type': {'db_refs': {'CL': 'CL:0000765'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '27827996',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -3373913262071139406,\n",
       "    'text': 'An increase in BRD3 occupancy at GATA1 sites following GATA1 activation is most significant as compared to that of BRD2 and BRD4 occupancy, suggesting that BRD3 recruitment is predominantly affected by GATA1 whereas BRD2 and BRD4 recruitment is regulated by factors besides GATA1 [XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS17111849',\n",
       "     'PII': 'E1849',\n",
       "     'PMCID': 'PMC5133849',\n",
       "     'PMID': '27827996',\n",
       "     'TRID': 16545836}},\n",
       "   {'annotations': {'agents': {'coords': [[31, 35], [47, 51]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '10090'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '29343972',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -1489016897268575299,\n",
       "    'text': 'Particularly, the BET proteins BRD2, BRD3, and BRD4 recognize and bind to the acetylated histone-lysine residues.',\n",
       "    'text_refs': {'DOI': '10.2147/OTT.S145971',\n",
       "     'PMCID': 'PMC5749389',\n",
       "     'PMID': '29343972',\n",
       "     'TRID': 28174205}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '27827996',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 1256873096803204907,\n",
       "    'text': 'BRD2 and BRD4 Interact with Viral Proteins and Contribute to Oncogenesis in Host Cells Infected with Viruses.',\n",
       "    'text_refs': {'DOI': '10.3390/IJMS17111849',\n",
       "     'PII': 'E1849',\n",
       "     'PMCID': 'PMC5133849',\n",
       "     'PMID': '27827996',\n",
       "     'TRID': 16545836}},\n",
       "   {'annotations': {'agents': {'coords': [[39, 43], [47, 51]]},\n",
       "     'found_by': 'binding9'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -5637226443740668066,\n",
       "    'text': 'JQ1 did not disrupt the interaction of BRD4 or BRD2 with LANA, but did reduce the binding of LANA with KSHV TR.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[31, 35], [40, 44]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'cell_line': {'db_refs': {'CELLOSAURUS': 'CVCL_9T72'},\n",
       "      'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12721'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 6328739575729990928,\n",
       "    'text': 'JQ1 is known to act rapidly on BRD2 and BRD4 binding to acetylated histones [XREF_BIBR], but its kinetic effects on transcription can be complex.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[29, 33], [37, 41]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'context': {'location': {'db_refs': {'GO': 'GO:0019015'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12721'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -4460096269124692310,\n",
       "    'text': 'To determine if JQ1 affected BRD4 or BRD2 interaction with LANA, we assayed the effects of JQ1 on the coimmunoprecipitation (coIP) of LANA with BRD4 (XREF_FIG, top panels) or with BRD2 (XREF_FIG, lower panels).',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[36, 40], [45, 49]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 1233503277108519757,\n",
       "    'text': 'We found that JQ1 treatment reduced BRD2 and BRD4 interactions at the TR regions by ~ 50% (XREF_FIG).',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[109, 113], [117, 121]]},\n",
       "     'found_by': 'binding9'},\n",
       "    'context': {'location': {'db_refs': {'GO': 'GO:0005694'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12721'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -5124102464736639245,\n",
       "    'text': 'JQ1 reduced binding of LANA to the TR and latency control region, but did not destabilize the interaction of BRD4 or BRD2 with LANA protein.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[46, 50], [52, 56]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -7561618092911824593,\n",
       "    'text': 'However, our finding that JQ1 rapidly reduces BRD2, BRD4, LANA and RAD21 interaction with KSHV genomes, and alters KSHV chromatin conformation independently of viral DNA replication, suggest that JQ1 can also act directly on the KSHV epigenome.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[18, 22], [27, 31]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -1726249009983314939,\n",
       "    'text': 'JQ1 disruption of BRD2 and BRD4 interaction with chromatin is known to occur at very early times, but how these early events regulate subsequent transcriptional and conformational events are not known.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[41, 45], [50, 54]]},\n",
       "     'found_by': 'binding7a'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 7926294519403353828,\n",
       "    'text': 'We suggest that JQ1 direct disruption of BRD2 and BRD4 interaction with KSHV chromatin leads to several subsequent events, including the loss of LANA binding to TR, loss of RAD21 dependent conformational control, and transcriptional derepression of KSHV lytic immediate early genes.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[79, 83], [87, 91]]},\n",
       "     'found_by': 'binding11'},\n",
       "    'context': {'location': {'db_refs': {'GO': 'GO:0019015'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '12721'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'hypothesis': True, 'section_type': None},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 8745564572145391722,\n",
       "    'text': 'However, we were unable to demonstrate any direct physical interaction between BRD2 or BRD4 with cohesin subunit RAD21 (XREF_SUPPLEMENTARY), suggesting the conformational control of KSHV latency involves additional factors.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'BIO-ASSOCIATE'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 2540504005750765779,\n",
       "    'text': 'We found that BRD2 and BRD4 associated with all three viral genome positions, showing the highest enrichment at the LANA binding sites at the TR.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -8275654358246284226,\n",
       "    'text': 'Neither BRD2 nor BRD4 bound to a region within the ORF37 gene (primer i), indicating that the enrichment at control regulatory regions is selective.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 3887134724099467843,\n",
       "    'text': 'We next asked whether JQ1 had any effect on the binding of BRD4 or BRD2 proteins to KSHV genome by ChIP assay ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -7506324149212461697,\n",
       "    'text': 'We found that JQ1 treatment reduced BRD2 and BRD4 interactions at the TR regions by ~50% ( xref ).',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': 8087079647002684130,\n",
       "    'text': 'These findings indicate that BRD2 and BRD4 can interact with viral regulatory regions, and that JQ1 can partially disrupt these interactions.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1153727522038570162,\n",
       "    'text': 'JQ1 disruption of BRD2 and BRD4 interaction with chromatin is known to occur at very early times, but how these early events regulate subsequent transcriptional and conformational events are not known.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'INTERACT'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -531056453524432072,\n",
       "    'text': 'We suggest that JQ1 direct disruption of BRD2 and BRD4 interaction with KSHV chromatin leads to several subsequent events, including the loss of LANA binding to TR, loss of RAD21-dependent conformational control, and transcriptional derepression of KSHV lytic immediate early genes.',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'found_by': 'BINDING'},\n",
       "    'pmid': '28107481',\n",
       "    'source_api': 'sparser',\n",
       "    'source_hash': -1448027423943845655,\n",
       "    'text': 'LANA is known to bind to BRD2 [ xref ] and BRD4 [ xref , xref ], and biophysical analyses revealed a direct interaction with the LANA DNA binding domain [ xref ].',\n",
       "    'text_refs': {'DOI': '10.1371/JOURNAL.PPAT.1006100',\n",
       "     'PMCID': 'PMC5287475',\n",
       "     'PMID': '28107481',\n",
       "     'TRID': 16352739}},\n",
       "   {'annotations': {'agents': {'coords': [[30, 34], [39, 43]]},\n",
       "     'found_by': 'binding1a'},\n",
       "    'context': {'location': {'db_refs': {'GO': 'GO:0005694'}, 'name': None},\n",
       "     'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '20709061',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': -6658758430223294719,\n",
       "    'text': 'Importantly, the BET proteins BRD2 and BRD4 associate with acetylated chromatin throughout the cell cycle [XREF_BIBR, XREF_BIBR], while other non BET bromodomain family members dissociate from the chromosomes during mitosis [XREF_BIBR].',\n",
       "    'text_refs': {'DOI': '10.1016/J.FEBSLET.2010.08.013',\n",
       "     'MANUSCRIPT_ID': 'NIHMS618856',\n",
       "     'PMCID': 'PMC4158924',\n",
       "     'PMID': '20709061',\n",
       "     'TRID': 1837339}},\n",
       "   {'annotations': {'agents': {'coords': [[82, 86], [94, 98]]},\n",
       "     'found_by': 'binding_of_and'},\n",
       "    'context': {'species': {'db_refs': {'TAXONOMY': '9606'}, 'name': None},\n",
       "     'type': 'bio'},\n",
       "    'epistemics': {'direct': True, 'section_type': None},\n",
       "    'pmid': '23604113',\n",
       "    'source_api': 'reach',\n",
       "    'source_hash': 448900016323780311,\n",
       "    'text': 'JQ1 is not specific for BRD-NUT, as it also competitively inhibits the binding of BRD2, BRD3, BRD4, and BRDT to chromatin, and NMC cells always have at least one intact BRD4 locus and express normal BRD4 as well as BRD-NUT oncoproteins; hence, the observed effects of JQ1 could stem from effects on BRD4, BRD-NUT, or both.',\n",
       "    'text_refs': {'DOI': '10.1038/ONC.2013.126',\n",
       "     'MANUSCRIPT_ID': 'NIHMS557273',\n",
       "     'PMCID': 'PMC3942361',\n",
       "     'PMID': '23604113',\n",
       "     'TRID': 14226131}}],\n",
       "  'id': 'fcf70adf-dc61-4999-b1ee-e17cbe0afb2b',\n",
       "  'matches_hash': '12484535149707124',\n",
       "  'members': [{'db_refs': {'EGID': '6046',\n",
       "     'HGNC': '1103',\n",
       "     'TEXT': 'BRD2',\n",
       "     'UP': 'P25440'},\n",
       "    'name': 'BRD2'},\n",
       "   {'db_refs': {'EGID': '23476',\n",
       "     'HGNC': '13575',\n",
       "     'TEXT': 'BRD4',\n",
       "     'UP': 'O60885'},\n",
       "    'name': 'BRD4'}],\n",
       "  'type': 'Complex'}]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0538e3c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "140"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(subnetwork_relations)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ffcf3f-ae8a-4612-9478-3b915d2f1ba6",
   "metadata": {},
   "source": [
    "## STEP 4: ASSEMBLE DATA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd43c119-0805-4079-a7ca-6bcf4ead9669",
   "metadata": {},
   "source": [
    "INDRA has a set of assembler classes to display the data for further analysis.  Below are some examples of assemblers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b9a3b1-6350-4366-83fa-6c467eca9df2",
   "metadata": {},
   "source": [
    "### HTML Assembler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6457b8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from indra.statements import stmts_from_json\n",
    "\n",
    "# Gather statistics for HTML presentation\n",
    "unique_stmts = {entry['data']['stmt_hash']: json.loads(entry['data']['stmt_json'])\n",
    "                for entry in subnetwork_relations}\n",
    "ev_counts_by_hash = {entry['data']['stmt_hash']: entry['data']['evidence_count']\n",
    "                     for entry in subnetwork_relations}\n",
    "source_counts_by_hash = {entry['data']['stmt_hash']: json.loads(entry['data']['source_counts'])\n",
    "                         for entry in subnetwork_relations}\n",
    "stmts = stmts_from_json(list(unique_stmts.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "5e1baaad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from indra.assemblers.html import HtmlAssembler\n",
    "ha = HtmlAssembler(stmts,\n",
    "                   title='INDRA subnetwork statements',\n",
    "                   db_rest_url='https://db.indra.bio',\n",
    "                   ev_counts=ev_counts_by_hash,\n",
    "                   source_counts=source_counts_by_hash)\n",
    "html_str = ha.make_model()\n",
    "ha.save_model('statements.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "70d66f7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.core.display import HTML\n",
    "# HTML(html_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3897b64c-da7f-4180-9ced-a46a3fca3c09",
   "metadata": {},
   "source": [
    "### NETWORK VISUALIZATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b88f2c71-1acd-4567-9068-3bd51b18898b",
   "metadata": {},
   "source": [
    "We can also visualize the subnetwork acquired from INDRA Cogex using an INDRA built-in assembler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7a0b5dee-1964-424c-9bbd-1238e550f2d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference: https://pygraphviz.github.io/documentation/stable/install.html\n",
    "!pip install -q pygraphviz \\\n",
    "    --config-settings=--global-option=build_ext \\\n",
    "    --config-settings=--global-option=\"-I$(brew --prefix graphviz)/include\" \\\n",
    "    --config-settings=--global-option=\"-L$(brew --prefix graphviz)/lib\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1988b07e-2863-4769-b41a-d1baf69a9725",
   "metadata": {},
   "outputs": [],
   "source": [
    "from indra.assemblers.graph.assembler import GraphAssembler\n",
    "\n",
    "ga = GraphAssembler(stmts=stmts)\n",
    "ga.make_model()\n",
    "ga.save_pdf(file_name='graph.pdf', prog='dot')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea640f1-0786-4f24-903f-d1553c65795e",
   "metadata": {},
   "source": [
    "### Tabular Format"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9777de80-a25d-4ddf-8408-4f0e57668331",
   "metadata": {},
   "source": [
    "INDRA also has an assembler for displaying INDRA cogex results as a table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "deaed566-a2ed-4ca1-9b29-f7925b51fa47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>agA_name</th>\n",
       "      <th>agB_name</th>\n",
       "      <th>agA_ns</th>\n",
       "      <th>agA_id</th>\n",
       "      <th>agB_ns</th>\n",
       "      <th>agB_id</th>\n",
       "      <th>residue</th>\n",
       "      <th>position</th>\n",
       "      <th>stmt_type</th>\n",
       "      <th>evidence_count</th>\n",
       "      <th>stmt_hash</th>\n",
       "      <th>belief</th>\n",
       "      <th>source_counts</th>\n",
       "      <th>initial_sign</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>BRD4</td>\n",
       "      <td>BRD2</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>13575</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1103</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>113</td>\n",
       "      <td>12484535149707124</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>{'reach': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>BRD2</td>\n",
       "      <td>BRD4</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1103</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>13575</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>113</td>\n",
       "      <td>12484535149707124</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>{'reach': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>BRD4</td>\n",
       "      <td>BRD3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>13575</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1104</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>51</td>\n",
       "      <td>359233681958482</td>\n",
       "      <td>0.999975</td>\n",
       "      <td>{'sparser': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>BRD3</td>\n",
       "      <td>BRD4</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1104</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>13575</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>51</td>\n",
       "      <td>359233681958482</td>\n",
       "      <td>0.999975</td>\n",
       "      <td>{'sparser': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>BRD2</td>\n",
       "      <td>BRD3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1103</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1104</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>42</td>\n",
       "      <td>-16515703320827288</td>\n",
       "      <td>0.999950</td>\n",
       "      <td>{'sparser': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>JMJD6</td>\n",
       "      <td>BRD2</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>19355</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1103</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>1</td>\n",
       "      <td>34194495135899326</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>{'sparser': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>JMJD6</td>\n",
       "      <td>BRD3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>19355</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1104</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>1</td>\n",
       "      <td>34194495135899326</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>{'sparser': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>KRR1</td>\n",
       "      <td>NOP2</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>5176</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>7867</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>1</td>\n",
       "      <td>-24130039704357981</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>{'biogrid': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>NOP2</td>\n",
       "      <td>KRR1</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>7867</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>5176</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Complex</td>\n",
       "      <td>1</td>\n",
       "      <td>-24130039704357981</td>\n",
       "      <td>0.980000</td>\n",
       "      <td>{'biogrid': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>BRD2</td>\n",
       "      <td>BRD3</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1103</td>\n",
       "      <td>HGNC</td>\n",
       "      <td>1104</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Inhibition</td>\n",
       "      <td>1</td>\n",
       "      <td>-20414889449388058</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>{'reach': 1}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    agA_name agB_name agA_ns agA_id agB_ns agB_id residue position  \\\n",
       "119     BRD4     BRD2   HGNC  13575   HGNC   1103    None     None   \n",
       "118     BRD2     BRD4   HGNC   1103   HGNC  13575    None     None   \n",
       "231     BRD4     BRD3   HGNC  13575   HGNC   1104    None     None   \n",
       "232     BRD3     BRD4   HGNC   1104   HGNC  13575    None     None   \n",
       "151     BRD2     BRD3   HGNC   1103   HGNC   1104    None     None   \n",
       "..       ...      ...    ...    ...    ...    ...     ...      ...   \n",
       "157    JMJD6     BRD2   HGNC  19355   HGNC   1103    None     None   \n",
       "158    JMJD6     BRD3   HGNC  19355   HGNC   1104    None     None   \n",
       "159     KRR1     NOP2   HGNC   5176   HGNC   7867    None     None   \n",
       "160     NOP2     KRR1   HGNC   7867   HGNC   5176    None     None   \n",
       "299     BRD2     BRD3   HGNC   1103   HGNC   1104    None     None   \n",
       "\n",
       "      stmt_type  evidence_count          stmt_hash    belief   source_counts  \\\n",
       "119     Complex             113  12484535149707124  0.999950    {'reach': 1}   \n",
       "118     Complex             113  12484535149707124  0.999950    {'reach': 1}   \n",
       "231     Complex              51    359233681958482  0.999975  {'sparser': 1}   \n",
       "232     Complex              51    359233681958482  0.999975  {'sparser': 1}   \n",
       "151     Complex              42 -16515703320827288  0.999950  {'sparser': 1}   \n",
       "..          ...             ...                ...       ...             ...   \n",
       "157     Complex               1  34194495135899326  0.650000  {'sparser': 1}   \n",
       "158     Complex               1  34194495135899326  0.650000  {'sparser': 1}   \n",
       "159     Complex               1 -24130039704357981  0.980000  {'biogrid': 1}   \n",
       "160     Complex               1 -24130039704357981  0.980000  {'biogrid': 1}   \n",
       "299  Inhibition               1 -20414889449388058  0.650000    {'reach': 1}   \n",
       "\n",
       "    initial_sign  \n",
       "119         None  \n",
       "118         None  \n",
       "231         None  \n",
       "232         None  \n",
       "151         None  \n",
       "..           ...  \n",
       "157         None  \n",
       "158         None  \n",
       "159         None  \n",
       "160         None  \n",
       "299         None  \n",
       "\n",
       "[300 rows x 14 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from indra.assemblers.indranet.assembler import IndraNetAssembler\n",
    "\n",
    "def add_evidence_column(stmt, ev_counts = ev_counts_by_hash):\n",
    "    hash = stmt.get_hash(refresh=True)\n",
    "    return ev_counts[hash]\n",
    "\n",
    "indra_net_assembler = IndraNetAssembler(statements=stmts)\n",
    "relations_table = indra_net_assembler.make_df(extra_columns=[('evidence_count', add_evidence_column)])\n",
    "relations_table = relations_table.sort_values(by=['evidence_count'], ascending=False)\n",
    "relations_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d3b7022",
   "metadata": {},
   "source": [
    "### NDEx network upload\n",
    "\n",
    "**This only works if you have a working NDEx account with credentials configured locally**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "3de02369",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating CX\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'https://www.ndexbio.org/v2/network/2eccaaf8-0ee1-11ef-9621-005056ae23aa'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ndex2.client\n",
    "from ndex2 import create_nice_cx_from_server\n",
    "from indra.assemblers.cx import NiceCxAssembler\n",
    "from indra.databases import ndex_client\n",
    "\n",
    "ca = NiceCxAssembler(stmts, 'INDRA May Institute Network')\n",
    "ncx = ca.make_model(self_loops=False)\n",
    "\n",
    "style_network_id = '058c452f-b0d6-11ea-a4d3-0660b7976219'\n",
    "style_ncx = create_nice_cx_from_server(\n",
    "    server='http://test.ndexbio.org',\n",
    "    uuid=style_network_id)\n",
    "ncx.apply_style_from_network(style_ncx)\n",
    "\n",
    "username, password = ndex_client.get_default_ndex_cred(ndex_cred=None)\n",
    "ndex_args = {'server': 'http://public.ndexbio.org',\n",
    "             'username': username,\n",
    "             'password': password}\n",
    "network_url = ncx.upload_to(**ndex_args)\n",
    "network_url"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc55c5df-a200-441a-a876-25ec7e2ea6f2",
   "metadata": {},
   "source": [
    "## ADDITIONAL STEP: MORE CREATIVE WAYS TO DISPLAY DATA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ee33ae-d819-4b3e-83d0-3d1876fa0fc1",
   "metadata": {},
   "source": [
    "### Example 1: Simple Network Visualization with P-values, Correlation, LogFC Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd63f84d-7591-4b52-a239-72f7ba2218f6",
   "metadata": {},
   "source": [
    "In this example, we take data from both MSstats functions `dataProcess` and `groupComparison`, gather metrics on p-value, correlation, and logFCs, and then contextualize that data with INDRA.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ed0bd6de-f277-495e-a226-f68563781aa5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RUN</th>\n",
       "      <th>Protein</th>\n",
       "      <th>LogIntensities</th>\n",
       "      <th>originalRUN</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SUBJECT</th>\n",
       "      <th>TotalGroupMeasurements</th>\n",
       "      <th>NumMeasuredFeature</th>\n",
       "      <th>MissingPercentage</th>\n",
       "      <th>more50missing</th>\n",
       "      <th>NumImputedFeature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1433B_HUMAN</td>\n",
       "      <td>12.873423</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate1_DMSO_A02_DIA</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>2</td>\n",
       "      <td>1210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1433B_HUMAN</td>\n",
       "      <td>12.866217</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate1_DMSO_A05_DIA</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>5</td>\n",
       "      <td>1210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1433B_HUMAN</td>\n",
       "      <td>12.686827</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate1_DMSO_A10_DIA</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>10</td>\n",
       "      <td>1210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>1433B_HUMAN</td>\n",
       "      <td>12.625462</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate1_DMSO_A12_DIA</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>12</td>\n",
       "      <td>1210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1433B_HUMAN</td>\n",
       "      <td>12.538365</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate1_DMSO_B01_DIA</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>13</td>\n",
       "      <td>1210</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189700</th>\n",
       "      <td>145</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>10.725469</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate3_PF477736_D05</td>\n",
       "      <td>DbET6</td>\n",
       "      <td>233</td>\n",
       "      <td>169</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189701</th>\n",
       "      <td>146</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>10.155338</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate3_DMSO_D06</td>\n",
       "      <td>DbET6</td>\n",
       "      <td>234</td>\n",
       "      <td>169</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189702</th>\n",
       "      <td>147</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>9.700678</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate3_K975_D12</td>\n",
       "      <td>DbET6</td>\n",
       "      <td>240</td>\n",
       "      <td>169</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189703</th>\n",
       "      <td>148</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>10.889323</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate3_VTP50469_F06</td>\n",
       "      <td>DbET6</td>\n",
       "      <td>258</td>\n",
       "      <td>169</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1189704</th>\n",
       "      <td>149</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>10.883954</td>\n",
       "      <td>230719_THP-1_Chrom_end2end_Plate3_K975_H01</td>\n",
       "      <td>DbET6</td>\n",
       "      <td>277</td>\n",
       "      <td>169</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>657552 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         RUN      Protein  LogIntensities  \\\n",
       "0          1  1433B_HUMAN       12.873423   \n",
       "1          2  1433B_HUMAN       12.866217   \n",
       "2          3  1433B_HUMAN       12.686827   \n",
       "3          4  1433B_HUMAN       12.625462   \n",
       "4          5  1433B_HUMAN       12.538365   \n",
       "...      ...          ...             ...   \n",
       "1189700  145   ZZZ3_HUMAN       10.725469   \n",
       "1189701  146   ZZZ3_HUMAN       10.155338   \n",
       "1189702  147   ZZZ3_HUMAN        9.700678   \n",
       "1189703  148   ZZZ3_HUMAN       10.889323   \n",
       "1189704  149   ZZZ3_HUMAN       10.883954   \n",
       "\n",
       "                                            originalRUN  GROUP  SUBJECT  \\\n",
       "0        230719_THP-1_Chrom_end2end_Plate1_DMSO_A02_DIA   DMSO        2   \n",
       "1        230719_THP-1_Chrom_end2end_Plate1_DMSO_A05_DIA   DMSO        5   \n",
       "2        230719_THP-1_Chrom_end2end_Plate1_DMSO_A10_DIA   DMSO       10   \n",
       "3        230719_THP-1_Chrom_end2end_Plate1_DMSO_A12_DIA   DMSO       12   \n",
       "4        230719_THP-1_Chrom_end2end_Plate1_DMSO_B01_DIA   DMSO       13   \n",
       "...                                                 ...    ...      ...   \n",
       "1189700  230719_THP-1_Chrom_end2end_Plate3_PF477736_D05  DbET6      233   \n",
       "1189701      230719_THP-1_Chrom_end2end_Plate3_DMSO_D06  DbET6      234   \n",
       "1189702      230719_THP-1_Chrom_end2end_Plate3_K975_D12  DbET6      240   \n",
       "1189703  230719_THP-1_Chrom_end2end_Plate3_VTP50469_F06  DbET6      258   \n",
       "1189704      230719_THP-1_Chrom_end2end_Plate3_K975_H01  DbET6      277   \n",
       "\n",
       "         TotalGroupMeasurements  NumMeasuredFeature  MissingPercentage  \\\n",
       "0                          1210                  10                0.0   \n",
       "1                          1210                  10                0.0   \n",
       "2                          1210                  10                0.0   \n",
       "3                          1210                  10                0.0   \n",
       "4                          1210                  10                0.0   \n",
       "...                         ...                 ...                ...   \n",
       "1189700                     169                  10                0.0   \n",
       "1189701                     169                  10                0.0   \n",
       "1189702                     169                  10                0.0   \n",
       "1189703                     169                  10                0.0   \n",
       "1189704                     169                  10                0.0   \n",
       "\n",
       "         more50missing  NumImputedFeature  \n",
       "0                False                  0  \n",
       "1                False                  0  \n",
       "2                False                  0  \n",
       "3                False                  0  \n",
       "4                False                  0  \n",
       "...                ...                ...  \n",
       "1189700          False                  0  \n",
       "1189701          False                  0  \n",
       "1189702          False                  0  \n",
       "1189703          False                  0  \n",
       "1189704          False                  0  \n",
       "\n",
       "[657552 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ABUNDANCE_PATH = \"../data/ProteinLevelData.csv\" # Set this path yourself\n",
    "ABUNDANCE_GROUPS_FILTER = ['DMSO', 'DbET6']\n",
    "def construct_abundance_df(filename):\n",
    "    pandas_df = pd.read_csv(filename)\n",
    "    pandas_df = pandas_df[pandas_df['GROUP'].isin(ABUNDANCE_GROUPS_FILTER)]\n",
    "    return pandas_df\n",
    "\n",
    "protein_abundance_df = construct_abundance_df(ABUNDANCE_PATH)\n",
    "protein_abundance_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1093a7c9-6e8a-43dc-bbe9-1591fbead34d",
   "metadata": {},
   "source": [
    "The `dataProcess` function outputs a dataset containing protein abundances for protein and biological replicate pair.  Using the protein abundances data, we can determine the correlation of abundance between pairs of proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f18e04b0-5a13-4f9c-aa53-5443a264f0d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BRD2_HUMAN</th>\n",
       "      <th>BRD3_HUMAN</th>\n",
       "      <th>BRD4_HUMAN</th>\n",
       "      <th>CEBPZ_HUMAN</th>\n",
       "      <th>CRNL1_HUMAN</th>\n",
       "      <th>DAZP1_HUMAN</th>\n",
       "      <th>FUBP2_HUMAN</th>\n",
       "      <th>FUBP3_HUMAN</th>\n",
       "      <th>GTPB4_HUMAN</th>\n",
       "      <th>HEAT3_HUMAN</th>\n",
       "      <th>...</th>\n",
       "      <th>KRR1_HUMAN</th>\n",
       "      <th>NOP2_HUMAN</th>\n",
       "      <th>PP1R8_HUMAN</th>\n",
       "      <th>PRP18_HUMAN</th>\n",
       "      <th>QKI_HUMAN</th>\n",
       "      <th>TIA1_HUMAN</th>\n",
       "      <th>UTP15_HUMAN</th>\n",
       "      <th>WDR12_HUMAN</th>\n",
       "      <th>WDR43_HUMAN</th>\n",
       "      <th>WDR75_HUMAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRD2_HUMAN</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.797708</td>\n",
       "      <td>0.879262</td>\n",
       "      <td>0.019850</td>\n",
       "      <td>-0.207924</td>\n",
       "      <td>0.598796</td>\n",
       "      <td>0.515316</td>\n",
       "      <td>0.557954</td>\n",
       "      <td>-0.093770</td>\n",
       "      <td>-0.429264</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.118297</td>\n",
       "      <td>-0.225047</td>\n",
       "      <td>0.621668</td>\n",
       "      <td>0.255623</td>\n",
       "      <td>0.479966</td>\n",
       "      <td>0.544447</td>\n",
       "      <td>-0.466654</td>\n",
       "      <td>-0.265250</td>\n",
       "      <td>-0.253804</td>\n",
       "      <td>-0.157479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRD3_HUMAN</th>\n",
       "      <td>0.797708</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.866766</td>\n",
       "      <td>-0.271567</td>\n",
       "      <td>-0.104151</td>\n",
       "      <td>0.312657</td>\n",
       "      <td>0.307522</td>\n",
       "      <td>0.226802</td>\n",
       "      <td>-0.387214</td>\n",
       "      <td>-0.175765</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.192796</td>\n",
       "      <td>-0.302674</td>\n",
       "      <td>0.422305</td>\n",
       "      <td>0.483052</td>\n",
       "      <td>0.660086</td>\n",
       "      <td>0.185024</td>\n",
       "      <td>-0.322859</td>\n",
       "      <td>-0.183338</td>\n",
       "      <td>-0.316502</td>\n",
       "      <td>-0.223405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRD4_HUMAN</th>\n",
       "      <td>0.879262</td>\n",
       "      <td>0.866766</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.027827</td>\n",
       "      <td>-0.155598</td>\n",
       "      <td>0.552971</td>\n",
       "      <td>0.435370</td>\n",
       "      <td>0.467304</td>\n",
       "      <td>-0.034894</td>\n",
       "      <td>-0.385975</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.068881</td>\n",
       "      <td>-0.274753</td>\n",
       "      <td>0.472644</td>\n",
       "      <td>0.275248</td>\n",
       "      <td>0.499226</td>\n",
       "      <td>0.489967</td>\n",
       "      <td>-0.423467</td>\n",
       "      <td>-0.236744</td>\n",
       "      <td>-0.388817</td>\n",
       "      <td>-0.146948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CEBPZ_HUMAN</th>\n",
       "      <td>0.019850</td>\n",
       "      <td>-0.271567</td>\n",
       "      <td>-0.027827</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.329054</td>\n",
       "      <td>0.367810</td>\n",
       "      <td>0.363832</td>\n",
       "      <td>0.413635</td>\n",
       "      <td>0.765247</td>\n",
       "      <td>-0.066058</td>\n",
       "      <td>...</td>\n",
       "      <td>0.575716</td>\n",
       "      <td>0.447506</td>\n",
       "      <td>0.250952</td>\n",
       "      <td>-0.262162</td>\n",
       "      <td>-0.105441</td>\n",
       "      <td>0.483017</td>\n",
       "      <td>0.208180</td>\n",
       "      <td>0.236918</td>\n",
       "      <td>0.353536</td>\n",
       "      <td>0.377506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CRNL1_HUMAN</th>\n",
       "      <td>-0.207924</td>\n",
       "      <td>-0.104151</td>\n",
       "      <td>-0.155598</td>\n",
       "      <td>0.329054</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.035805</td>\n",
       "      <td>0.107018</td>\n",
       "      <td>0.026768</td>\n",
       "      <td>0.230088</td>\n",
       "      <td>0.231355</td>\n",
       "      <td>...</td>\n",
       "      <td>0.396803</td>\n",
       "      <td>0.527981</td>\n",
       "      <td>-0.047657</td>\n",
       "      <td>0.219325</td>\n",
       "      <td>0.151090</td>\n",
       "      <td>-0.183671</td>\n",
       "      <td>0.600379</td>\n",
       "      <td>0.558789</td>\n",
       "      <td>0.566930</td>\n",
       "      <td>0.510267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DAZP1_HUMAN</th>\n",
       "      <td>0.598796</td>\n",
       "      <td>0.312657</td>\n",
       "      <td>0.552971</td>\n",
       "      <td>0.367810</td>\n",
       "      <td>-0.035805</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.608002</td>\n",
       "      <td>0.743681</td>\n",
       "      <td>0.329881</td>\n",
       "      <td>-0.311158</td>\n",
       "      <td>...</td>\n",
       "      <td>0.199983</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>0.595334</td>\n",
       "      <td>0.062600</td>\n",
       "      <td>0.372729</td>\n",
       "      <td>0.751419</td>\n",
       "      <td>-0.268644</td>\n",
       "      <td>-0.070602</td>\n",
       "      <td>-0.132739</td>\n",
       "      <td>0.042042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FUBP2_HUMAN</th>\n",
       "      <td>0.515316</td>\n",
       "      <td>0.307522</td>\n",
       "      <td>0.435370</td>\n",
       "      <td>0.363832</td>\n",
       "      <td>0.107018</td>\n",
       "      <td>0.608002</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.610191</td>\n",
       "      <td>0.168864</td>\n",
       "      <td>-0.218615</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282158</td>\n",
       "      <td>0.194035</td>\n",
       "      <td>0.513343</td>\n",
       "      <td>0.237327</td>\n",
       "      <td>0.469106</td>\n",
       "      <td>0.589631</td>\n",
       "      <td>-0.048102</td>\n",
       "      <td>0.109012</td>\n",
       "      <td>0.165000</td>\n",
       "      <td>0.105673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FUBP3_HUMAN</th>\n",
       "      <td>0.557954</td>\n",
       "      <td>0.226802</td>\n",
       "      <td>0.467304</td>\n",
       "      <td>0.413635</td>\n",
       "      <td>0.026768</td>\n",
       "      <td>0.743681</td>\n",
       "      <td>0.610191</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.419721</td>\n",
       "      <td>-0.322835</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156421</td>\n",
       "      <td>0.070900</td>\n",
       "      <td>0.604436</td>\n",
       "      <td>0.042284</td>\n",
       "      <td>0.303556</td>\n",
       "      <td>0.790524</td>\n",
       "      <td>-0.295023</td>\n",
       "      <td>-0.032122</td>\n",
       "      <td>0.020199</td>\n",
       "      <td>0.132255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTPB4_HUMAN</th>\n",
       "      <td>-0.093770</td>\n",
       "      <td>-0.387214</td>\n",
       "      <td>-0.034894</td>\n",
       "      <td>0.765247</td>\n",
       "      <td>0.230088</td>\n",
       "      <td>0.329881</td>\n",
       "      <td>0.168864</td>\n",
       "      <td>0.419721</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.044000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.484970</td>\n",
       "      <td>0.288195</td>\n",
       "      <td>0.112460</td>\n",
       "      <td>-0.407856</td>\n",
       "      <td>-0.330488</td>\n",
       "      <td>0.534385</td>\n",
       "      <td>0.066697</td>\n",
       "      <td>0.030891</td>\n",
       "      <td>0.189041</td>\n",
       "      <td>0.265675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HEAT3_HUMAN</th>\n",
       "      <td>-0.429264</td>\n",
       "      <td>-0.175765</td>\n",
       "      <td>-0.385975</td>\n",
       "      <td>-0.066058</td>\n",
       "      <td>0.231355</td>\n",
       "      <td>-0.311158</td>\n",
       "      <td>-0.218615</td>\n",
       "      <td>-0.322835</td>\n",
       "      <td>-0.044000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.199914</td>\n",
       "      <td>0.096674</td>\n",
       "      <td>-0.343605</td>\n",
       "      <td>0.141172</td>\n",
       "      <td>-0.025871</td>\n",
       "      <td>-0.320925</td>\n",
       "      <td>0.445552</td>\n",
       "      <td>0.254046</td>\n",
       "      <td>0.246794</td>\n",
       "      <td>-0.046507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HNRPD_HUMAN</th>\n",
       "      <td>0.539103</td>\n",
       "      <td>0.340189</td>\n",
       "      <td>0.351521</td>\n",
       "      <td>0.113002</td>\n",
       "      <td>0.014405</td>\n",
       "      <td>0.545040</td>\n",
       "      <td>0.584445</td>\n",
       "      <td>0.488188</td>\n",
       "      <td>-0.114297</td>\n",
       "      <td>-0.274164</td>\n",
       "      <td>...</td>\n",
       "      <td>0.125444</td>\n",
       "      <td>0.092278</td>\n",
       "      <td>0.563730</td>\n",
       "      <td>0.317165</td>\n",
       "      <td>0.557362</td>\n",
       "      <td>0.428906</td>\n",
       "      <td>-0.083049</td>\n",
       "      <td>0.235416</td>\n",
       "      <td>0.175258</td>\n",
       "      <td>0.073476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRR1_HUMAN</th>\n",
       "      <td>-0.118297</td>\n",
       "      <td>-0.192796</td>\n",
       "      <td>-0.068881</td>\n",
       "      <td>0.575716</td>\n",
       "      <td>0.396803</td>\n",
       "      <td>0.199983</td>\n",
       "      <td>0.282158</td>\n",
       "      <td>0.156421</td>\n",
       "      <td>0.484970</td>\n",
       "      <td>0.199914</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.398750</td>\n",
       "      <td>0.056222</td>\n",
       "      <td>-0.123099</td>\n",
       "      <td>-0.017384</td>\n",
       "      <td>0.231466</td>\n",
       "      <td>0.421255</td>\n",
       "      <td>0.263342</td>\n",
       "      <td>0.412868</td>\n",
       "      <td>0.286653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOP2_HUMAN</th>\n",
       "      <td>-0.225047</td>\n",
       "      <td>-0.302674</td>\n",
       "      <td>-0.274753</td>\n",
       "      <td>0.447506</td>\n",
       "      <td>0.527981</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>0.194035</td>\n",
       "      <td>0.070900</td>\n",
       "      <td>0.288195</td>\n",
       "      <td>0.096674</td>\n",
       "      <td>...</td>\n",
       "      <td>0.398750</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.054507</td>\n",
       "      <td>0.052304</td>\n",
       "      <td>-0.065705</td>\n",
       "      <td>-0.018197</td>\n",
       "      <td>0.488759</td>\n",
       "      <td>0.491605</td>\n",
       "      <td>0.588624</td>\n",
       "      <td>0.490429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PP1R8_HUMAN</th>\n",
       "      <td>0.621668</td>\n",
       "      <td>0.422305</td>\n",
       "      <td>0.472644</td>\n",
       "      <td>0.250952</td>\n",
       "      <td>-0.047657</td>\n",
       "      <td>0.595334</td>\n",
       "      <td>0.513343</td>\n",
       "      <td>0.604436</td>\n",
       "      <td>0.112460</td>\n",
       "      <td>-0.343605</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056222</td>\n",
       "      <td>0.054507</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.204392</td>\n",
       "      <td>0.499983</td>\n",
       "      <td>0.548012</td>\n",
       "      <td>-0.286920</td>\n",
       "      <td>0.039424</td>\n",
       "      <td>0.045308</td>\n",
       "      <td>0.149858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRP18_HUMAN</th>\n",
       "      <td>0.255623</td>\n",
       "      <td>0.483052</td>\n",
       "      <td>0.275248</td>\n",
       "      <td>-0.262162</td>\n",
       "      <td>0.219325</td>\n",
       "      <td>0.062600</td>\n",
       "      <td>0.237327</td>\n",
       "      <td>0.042284</td>\n",
       "      <td>-0.407856</td>\n",
       "      <td>0.141172</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.123099</td>\n",
       "      <td>0.052304</td>\n",
       "      <td>0.204392</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.599861</td>\n",
       "      <td>-0.097829</td>\n",
       "      <td>0.117953</td>\n",
       "      <td>0.305236</td>\n",
       "      <td>0.109525</td>\n",
       "      <td>-0.033367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QKI_HUMAN</th>\n",
       "      <td>0.479966</td>\n",
       "      <td>0.660086</td>\n",
       "      <td>0.499226</td>\n",
       "      <td>-0.105441</td>\n",
       "      <td>0.151090</td>\n",
       "      <td>0.372729</td>\n",
       "      <td>0.469106</td>\n",
       "      <td>0.303556</td>\n",
       "      <td>-0.330488</td>\n",
       "      <td>-0.025871</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.017384</td>\n",
       "      <td>-0.065705</td>\n",
       "      <td>0.499983</td>\n",
       "      <td>0.599861</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.169855</td>\n",
       "      <td>-0.017787</td>\n",
       "      <td>0.226580</td>\n",
       "      <td>0.060119</td>\n",
       "      <td>0.018128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TIA1_HUMAN</th>\n",
       "      <td>0.544447</td>\n",
       "      <td>0.185024</td>\n",
       "      <td>0.489967</td>\n",
       "      <td>0.483017</td>\n",
       "      <td>-0.183671</td>\n",
       "      <td>0.751419</td>\n",
       "      <td>0.589631</td>\n",
       "      <td>0.790524</td>\n",
       "      <td>0.534385</td>\n",
       "      <td>-0.320925</td>\n",
       "      <td>...</td>\n",
       "      <td>0.231466</td>\n",
       "      <td>-0.018197</td>\n",
       "      <td>0.548012</td>\n",
       "      <td>-0.097829</td>\n",
       "      <td>0.169855</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.386969</td>\n",
       "      <td>-0.180420</td>\n",
       "      <td>-0.050802</td>\n",
       "      <td>-0.023821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTP15_HUMAN</th>\n",
       "      <td>-0.466654</td>\n",
       "      <td>-0.322859</td>\n",
       "      <td>-0.423467</td>\n",
       "      <td>0.208180</td>\n",
       "      <td>0.600379</td>\n",
       "      <td>-0.268644</td>\n",
       "      <td>-0.048102</td>\n",
       "      <td>-0.295023</td>\n",
       "      <td>0.066697</td>\n",
       "      <td>0.445552</td>\n",
       "      <td>...</td>\n",
       "      <td>0.421255</td>\n",
       "      <td>0.488759</td>\n",
       "      <td>-0.286920</td>\n",
       "      <td>0.117953</td>\n",
       "      <td>-0.017787</td>\n",
       "      <td>-0.386969</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.462224</td>\n",
       "      <td>0.584313</td>\n",
       "      <td>0.331612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR12_HUMAN</th>\n",
       "      <td>-0.265250</td>\n",
       "      <td>-0.183338</td>\n",
       "      <td>-0.236744</td>\n",
       "      <td>0.236918</td>\n",
       "      <td>0.558789</td>\n",
       "      <td>-0.070602</td>\n",
       "      <td>0.109012</td>\n",
       "      <td>-0.032122</td>\n",
       "      <td>0.030891</td>\n",
       "      <td>0.254046</td>\n",
       "      <td>...</td>\n",
       "      <td>0.263342</td>\n",
       "      <td>0.491605</td>\n",
       "      <td>0.039424</td>\n",
       "      <td>0.305236</td>\n",
       "      <td>0.226580</td>\n",
       "      <td>-0.180420</td>\n",
       "      <td>0.462224</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.519296</td>\n",
       "      <td>0.411982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR43_HUMAN</th>\n",
       "      <td>-0.253804</td>\n",
       "      <td>-0.316502</td>\n",
       "      <td>-0.388817</td>\n",
       "      <td>0.353536</td>\n",
       "      <td>0.566930</td>\n",
       "      <td>-0.132739</td>\n",
       "      <td>0.165000</td>\n",
       "      <td>0.020199</td>\n",
       "      <td>0.189041</td>\n",
       "      <td>0.246794</td>\n",
       "      <td>...</td>\n",
       "      <td>0.412868</td>\n",
       "      <td>0.588624</td>\n",
       "      <td>0.045308</td>\n",
       "      <td>0.109525</td>\n",
       "      <td>0.060119</td>\n",
       "      <td>-0.050802</td>\n",
       "      <td>0.584313</td>\n",
       "      <td>0.519296</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.437904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WDR75_HUMAN</th>\n",
       "      <td>-0.157479</td>\n",
       "      <td>-0.223405</td>\n",
       "      <td>-0.146948</td>\n",
       "      <td>0.377506</td>\n",
       "      <td>0.510267</td>\n",
       "      <td>0.042042</td>\n",
       "      <td>0.105673</td>\n",
       "      <td>0.132255</td>\n",
       "      <td>0.265675</td>\n",
       "      <td>-0.046507</td>\n",
       "      <td>...</td>\n",
       "      <td>0.286653</td>\n",
       "      <td>0.490429</td>\n",
       "      <td>0.149858</td>\n",
       "      <td>-0.033367</td>\n",
       "      <td>0.018128</td>\n",
       "      <td>-0.023821</td>\n",
       "      <td>0.331612</td>\n",
       "      <td>0.411982</td>\n",
       "      <td>0.437904</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             BRD2_HUMAN  BRD3_HUMAN  BRD4_HUMAN  CEBPZ_HUMAN  CRNL1_HUMAN  \\\n",
       "BRD2_HUMAN     1.000000    0.797708    0.879262     0.019850    -0.207924   \n",
       "BRD3_HUMAN     0.797708    1.000000    0.866766    -0.271567    -0.104151   \n",
       "BRD4_HUMAN     0.879262    0.866766    1.000000    -0.027827    -0.155598   \n",
       "CEBPZ_HUMAN    0.019850   -0.271567   -0.027827     1.000000     0.329054   \n",
       "CRNL1_HUMAN   -0.207924   -0.104151   -0.155598     0.329054     1.000000   \n",
       "DAZP1_HUMAN    0.598796    0.312657    0.552971     0.367810    -0.035805   \n",
       "FUBP2_HUMAN    0.515316    0.307522    0.435370     0.363832     0.107018   \n",
       "FUBP3_HUMAN    0.557954    0.226802    0.467304     0.413635     0.026768   \n",
       "GTPB4_HUMAN   -0.093770   -0.387214   -0.034894     0.765247     0.230088   \n",
       "HEAT3_HUMAN   -0.429264   -0.175765   -0.385975    -0.066058     0.231355   \n",
       "HNRPD_HUMAN    0.539103    0.340189    0.351521     0.113002     0.014405   \n",
       "KRR1_HUMAN    -0.118297   -0.192796   -0.068881     0.575716     0.396803   \n",
       "NOP2_HUMAN    -0.225047   -0.302674   -0.274753     0.447506     0.527981   \n",
       "PP1R8_HUMAN    0.621668    0.422305    0.472644     0.250952    -0.047657   \n",
       "PRP18_HUMAN    0.255623    0.483052    0.275248    -0.262162     0.219325   \n",
       "QKI_HUMAN      0.479966    0.660086    0.499226    -0.105441     0.151090   \n",
       "TIA1_HUMAN     0.544447    0.185024    0.489967     0.483017    -0.183671   \n",
       "UTP15_HUMAN   -0.466654   -0.322859   -0.423467     0.208180     0.600379   \n",
       "WDR12_HUMAN   -0.265250   -0.183338   -0.236744     0.236918     0.558789   \n",
       "WDR43_HUMAN   -0.253804   -0.316502   -0.388817     0.353536     0.566930   \n",
       "WDR75_HUMAN   -0.157479   -0.223405   -0.146948     0.377506     0.510267   \n",
       "\n",
       "             DAZP1_HUMAN  FUBP2_HUMAN  FUBP3_HUMAN  GTPB4_HUMAN  HEAT3_HUMAN  \\\n",
       "BRD2_HUMAN      0.598796     0.515316     0.557954    -0.093770    -0.429264   \n",
       "BRD3_HUMAN      0.312657     0.307522     0.226802    -0.387214    -0.175765   \n",
       "BRD4_HUMAN      0.552971     0.435370     0.467304    -0.034894    -0.385975   \n",
       "CEBPZ_HUMAN     0.367810     0.363832     0.413635     0.765247    -0.066058   \n",
       "CRNL1_HUMAN    -0.035805     0.107018     0.026768     0.230088     0.231355   \n",
       "DAZP1_HUMAN     1.000000     0.608002     0.743681     0.329881    -0.311158   \n",
       "FUBP2_HUMAN     0.608002     1.000000     0.610191     0.168864    -0.218615   \n",
       "FUBP3_HUMAN     0.743681     0.610191     1.000000     0.419721    -0.322835   \n",
       "GTPB4_HUMAN     0.329881     0.168864     0.419721     1.000000    -0.044000   \n",
       "HEAT3_HUMAN    -0.311158    -0.218615    -0.322835    -0.044000     1.000000   \n",
       "HNRPD_HUMAN     0.545040     0.584445     0.488188    -0.114297    -0.274164   \n",
       "KRR1_HUMAN      0.199983     0.282158     0.156421     0.484970     0.199914   \n",
       "NOP2_HUMAN      0.006942     0.194035     0.070900     0.288195     0.096674   \n",
       "PP1R8_HUMAN     0.595334     0.513343     0.604436     0.112460    -0.343605   \n",
       "PRP18_HUMAN     0.062600     0.237327     0.042284    -0.407856     0.141172   \n",
       "QKI_HUMAN       0.372729     0.469106     0.303556    -0.330488    -0.025871   \n",
       "TIA1_HUMAN      0.751419     0.589631     0.790524     0.534385    -0.320925   \n",
       "UTP15_HUMAN    -0.268644    -0.048102    -0.295023     0.066697     0.445552   \n",
       "WDR12_HUMAN    -0.070602     0.109012    -0.032122     0.030891     0.254046   \n",
       "WDR43_HUMAN    -0.132739     0.165000     0.020199     0.189041     0.246794   \n",
       "WDR75_HUMAN     0.042042     0.105673     0.132255     0.265675    -0.046507   \n",
       "\n",
       "             ...  KRR1_HUMAN  NOP2_HUMAN  PP1R8_HUMAN  PRP18_HUMAN  QKI_HUMAN  \\\n",
       "BRD2_HUMAN   ...   -0.118297   -0.225047     0.621668     0.255623   0.479966   \n",
       "BRD3_HUMAN   ...   -0.192796   -0.302674     0.422305     0.483052   0.660086   \n",
       "BRD4_HUMAN   ...   -0.068881   -0.274753     0.472644     0.275248   0.499226   \n",
       "CEBPZ_HUMAN  ...    0.575716    0.447506     0.250952    -0.262162  -0.105441   \n",
       "CRNL1_HUMAN  ...    0.396803    0.527981    -0.047657     0.219325   0.151090   \n",
       "DAZP1_HUMAN  ...    0.199983    0.006942     0.595334     0.062600   0.372729   \n",
       "FUBP2_HUMAN  ...    0.282158    0.194035     0.513343     0.237327   0.469106   \n",
       "FUBP3_HUMAN  ...    0.156421    0.070900     0.604436     0.042284   0.303556   \n",
       "GTPB4_HUMAN  ...    0.484970    0.288195     0.112460    -0.407856  -0.330488   \n",
       "HEAT3_HUMAN  ...    0.199914    0.096674    -0.343605     0.141172  -0.025871   \n",
       "HNRPD_HUMAN  ...    0.125444    0.092278     0.563730     0.317165   0.557362   \n",
       "KRR1_HUMAN   ...    1.000000    0.398750     0.056222    -0.123099  -0.017384   \n",
       "NOP2_HUMAN   ...    0.398750    1.000000     0.054507     0.052304  -0.065705   \n",
       "PP1R8_HUMAN  ...    0.056222    0.054507     1.000000     0.204392   0.499983   \n",
       "PRP18_HUMAN  ...   -0.123099    0.052304     0.204392     1.000000   0.599861   \n",
       "QKI_HUMAN    ...   -0.017384   -0.065705     0.499983     0.599861   1.000000   \n",
       "TIA1_HUMAN   ...    0.231466   -0.018197     0.548012    -0.097829   0.169855   \n",
       "UTP15_HUMAN  ...    0.421255    0.488759    -0.286920     0.117953  -0.017787   \n",
       "WDR12_HUMAN  ...    0.263342    0.491605     0.039424     0.305236   0.226580   \n",
       "WDR43_HUMAN  ...    0.412868    0.588624     0.045308     0.109525   0.060119   \n",
       "WDR75_HUMAN  ...    0.286653    0.490429     0.149858    -0.033367   0.018128   \n",
       "\n",
       "             TIA1_HUMAN  UTP15_HUMAN  WDR12_HUMAN  WDR43_HUMAN  WDR75_HUMAN  \n",
       "BRD2_HUMAN     0.544447    -0.466654    -0.265250    -0.253804    -0.157479  \n",
       "BRD3_HUMAN     0.185024    -0.322859    -0.183338    -0.316502    -0.223405  \n",
       "BRD4_HUMAN     0.489967    -0.423467    -0.236744    -0.388817    -0.146948  \n",
       "CEBPZ_HUMAN    0.483017     0.208180     0.236918     0.353536     0.377506  \n",
       "CRNL1_HUMAN   -0.183671     0.600379     0.558789     0.566930     0.510267  \n",
       "DAZP1_HUMAN    0.751419    -0.268644    -0.070602    -0.132739     0.042042  \n",
       "FUBP2_HUMAN    0.589631    -0.048102     0.109012     0.165000     0.105673  \n",
       "FUBP3_HUMAN    0.790524    -0.295023    -0.032122     0.020199     0.132255  \n",
       "GTPB4_HUMAN    0.534385     0.066697     0.030891     0.189041     0.265675  \n",
       "HEAT3_HUMAN   -0.320925     0.445552     0.254046     0.246794    -0.046507  \n",
       "HNRPD_HUMAN    0.428906    -0.083049     0.235416     0.175258     0.073476  \n",
       "KRR1_HUMAN     0.231466     0.421255     0.263342     0.412868     0.286653  \n",
       "NOP2_HUMAN    -0.018197     0.488759     0.491605     0.588624     0.490429  \n",
       "PP1R8_HUMAN    0.548012    -0.286920     0.039424     0.045308     0.149858  \n",
       "PRP18_HUMAN   -0.097829     0.117953     0.305236     0.109525    -0.033367  \n",
       "QKI_HUMAN      0.169855    -0.017787     0.226580     0.060119     0.018128  \n",
       "TIA1_HUMAN     1.000000    -0.386969    -0.180420    -0.050802    -0.023821  \n",
       "UTP15_HUMAN   -0.386969     1.000000     0.462224     0.584313     0.331612  \n",
       "WDR12_HUMAN   -0.180420     0.462224     1.000000     0.519296     0.411982  \n",
       "WDR43_HUMAN   -0.050802     0.584313     0.519296     1.000000     0.437904  \n",
       "WDR75_HUMAN   -0.023821     0.331612     0.411982     0.437904     1.000000  \n",
       "\n",
       "[21 rows x 21 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "def calculate_correlation_matrix(pvalue_df, protein_level_summary):\n",
    "    data = {}\n",
    "    subjects = protein_level_summary['SUBJECT'].unique()\n",
    "    for protein in pvalue_df['Protein'].unique():\n",
    "        data[protein] = []\n",
    "        protein_level_df = protein_level_summary[protein_level_summary['Protein'] == protein]\n",
    "        for subject in subjects:\n",
    "            if subject in protein_level_df['SUBJECT'].values:\n",
    "                protein_level_df_subject = protein_level_df[protein_level_df['SUBJECT'] == subject]\n",
    "                data[protein].append(protein_level_df_subject['LogIntensities'].iloc[0])\n",
    "            else:\n",
    "                data[protein].append(np.nan)\n",
    "    df = pd.DataFrame(data)\n",
    "    corrM = df.corr() \n",
    "    return corrM\n",
    "\n",
    "corr_matrix = calculate_correlation_matrix(pvalue_logfc_df, protein_abundance_df)\n",
    "corr_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8a24246e-a8d0-4be5-8fbf-76d36b3eab77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Label</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>SE</th>\n",
       "      <th>Tvalue</th>\n",
       "      <th>DF</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>adj.pvalue</th>\n",
       "      <th>issue</th>\n",
       "      <th>MissingPercentage</th>\n",
       "      <th>ImputationPercentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19890</th>\n",
       "      <td>BRD2_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>2.046185</td>\n",
       "      <td>0.114339</td>\n",
       "      <td>17.895836</td>\n",
       "      <td>260.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.310067</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19935</th>\n",
       "      <td>BRD3_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>3.333428</td>\n",
       "      <td>0.126571</td>\n",
       "      <td>26.336522</td>\n",
       "      <td>257.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.252349</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19980</th>\n",
       "      <td>BRD4_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>2.668935</td>\n",
       "      <td>0.101283</td>\n",
       "      <td>26.351317</td>\n",
       "      <td>257.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.118121</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41580</th>\n",
       "      <td>DAZP1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.617508</td>\n",
       "      <td>0.099537</td>\n",
       "      <td>6.203818</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.159950e-09</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.286577</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66060</th>\n",
       "      <td>FUBP2_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.291045</td>\n",
       "      <td>0.077226</td>\n",
       "      <td>3.768747</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.030530e-04</td>\n",
       "      <td>0.047824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017450</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66105</th>\n",
       "      <td>FUBP3_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.300410</td>\n",
       "      <td>0.069882</td>\n",
       "      <td>4.298798</td>\n",
       "      <td>260.0</td>\n",
       "      <td>2.430235e-05</td>\n",
       "      <td>0.008941</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.095302</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137745</th>\n",
       "      <td>PP1R8_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.358282</td>\n",
       "      <td>0.069056</td>\n",
       "      <td>5.188314</td>\n",
       "      <td>260.0</td>\n",
       "      <td>4.275951e-07</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.045638</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140535</th>\n",
       "      <td>PRP18_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.485713</td>\n",
       "      <td>0.120799</td>\n",
       "      <td>4.020834</td>\n",
       "      <td>260.0</td>\n",
       "      <td>7.603588e-05</td>\n",
       "      <td>0.022122</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.049664</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145440</th>\n",
       "      <td>QKI_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.336985</td>\n",
       "      <td>0.055130</td>\n",
       "      <td>6.112567</td>\n",
       "      <td>260.0</td>\n",
       "      <td>3.570594e-09</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196155</th>\n",
       "      <td>TIA1_HUMAN</td>\n",
       "      <td>DMSO-DbET6</td>\n",
       "      <td>0.476791</td>\n",
       "      <td>0.105361</td>\n",
       "      <td>4.525306</td>\n",
       "      <td>260.0</td>\n",
       "      <td>9.182867e-06</td>\n",
       "      <td>0.004129</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.058389</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Protein       Label    log2FC        SE     Tvalue     DF  \\\n",
       "19890    BRD2_HUMAN  DMSO-DbET6  2.046185  0.114339  17.895836  260.0   \n",
       "19935    BRD3_HUMAN  DMSO-DbET6  3.333428  0.126571  26.336522  257.0   \n",
       "19980    BRD4_HUMAN  DMSO-DbET6  2.668935  0.101283  26.351317  257.0   \n",
       "41580   DAZP1_HUMAN  DMSO-DbET6  0.617508  0.099537   6.203818  260.0   \n",
       "66060   FUBP2_HUMAN  DMSO-DbET6  0.291045  0.077226   3.768747  260.0   \n",
       "66105   FUBP3_HUMAN  DMSO-DbET6  0.300410  0.069882   4.298798  260.0   \n",
       "137745  PP1R8_HUMAN  DMSO-DbET6  0.358282  0.069056   5.188314  260.0   \n",
       "140535  PRP18_HUMAN  DMSO-DbET6  0.485713  0.120799   4.020834  260.0   \n",
       "145440    QKI_HUMAN  DMSO-DbET6  0.336985  0.055130   6.112567  260.0   \n",
       "196155   TIA1_HUMAN  DMSO-DbET6  0.476791  0.105361   4.525306  260.0   \n",
       "\n",
       "              pvalue  adj.pvalue issue  MissingPercentage  \\\n",
       "19890   0.000000e+00    0.000000   NaN           0.310067   \n",
       "19935   0.000000e+00    0.000000   NaN           0.252349   \n",
       "19980   0.000000e+00    0.000000   NaN           0.118121   \n",
       "41580   2.159950e-09    0.000002   NaN           0.286577   \n",
       "66060   2.030530e-04    0.047824   NaN           0.017450   \n",
       "66105   2.430235e-05    0.008941   NaN           0.095302   \n",
       "137745  4.275951e-07    0.000235   NaN           0.045638   \n",
       "140535  7.603588e-05    0.022122   NaN           0.049664   \n",
       "145440  3.570594e-09    0.000003   NaN           0.000000   \n",
       "196155  9.182867e-06    0.004129   NaN           0.058389   \n",
       "\n",
       "        ImputationPercentage  \n",
       "19890                    0.0  \n",
       "19935                    0.0  \n",
       "19980                    0.0  \n",
       "41580                    0.0  \n",
       "66060                    0.0  \n",
       "66105                    0.0  \n",
       "137745                   0.0  \n",
       "140535                   0.0  \n",
       "145440                   0.0  \n",
       "196155                   0.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LOG_FC_FILTER = 0.25\n",
    "pvalue_logfc_df['log2FC'] = pvalue_logfc_df['log2FC'].astype(float)\n",
    "logfc_proteins = pvalue_logfc_df[pvalue_logfc_df['log2FC'] > LOG_FC_FILTER]\n",
    "logfc_proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fcebda6d-3eeb-4359-93e1-ea0f5e361522",
   "metadata": {},
   "outputs": [],
   "source": [
    "CORRELATION_FILTER = 0.3\n",
    "EVIDENCE_FILTER = 10\n",
    "\n",
    "for node in ga.graph.nodes():\n",
    "    node_properties = {\n",
    "        'color': '#808080',\n",
    "        'shape': 'Mrecord',\n",
    "        'fontsize': 8\n",
    "    }\n",
    "    protein_id = hgnc_gene_name_to_uniprot_mapping.get(node)\n",
    "    if not protein_id:\n",
    "        ga.graph.add_node(node,\n",
    "                          label=node,\n",
    "                          **node_properties)\n",
    "    elif protein_id in logfc_proteins['Protein'].values:\n",
    "        node_properties['color'] = '#00FF00'\n",
    "        logFC_value = round(pvalue_logfc_df[pvalue_logfc_df['Protein'] == protein_id]['log2FC'].iloc[0], 2)\n",
    "        ga.graph.add_node(node,\n",
    "                          label=f'{node}: {logFC_value} LogFC',\n",
    "                          **node_properties)\n",
    "    else:\n",
    "        node_properties['color'] = '#FF0000'\n",
    "        logFC_value = round(pvalue_logfc_df[pvalue_logfc_df['Protein'] == protein_id]['log2FC'].iloc[0], 2)\n",
    "        ga.graph.add_node(node,\n",
    "                          label=f'{node}: {logFC_value} LogFC',\n",
    "                          **node_properties)\n",
    "\n",
    "color = '#ff0000'\n",
    "color_default = '#000000'\n",
    "for edge in ga.graph.edges():\n",
    "    params = {'color': color_default,\n",
    "              'arrowhead': 'normal',\n",
    "              'dir': 'forward'}\n",
    "    if edge[0] in hgnc_gene_name_to_uniprot_mapping.keys() and \\\n",
    "        edge[1] in hgnc_gene_name_to_uniprot_mapping.keys():\n",
    "        uniprot_0 = hgnc_gene_name_to_uniprot_mapping.get(edge[0])\n",
    "        uniprot_1 = hgnc_gene_name_to_uniprot_mapping.get(edge[1])\n",
    "        correlation = round(corr_matrix[uniprot_0][uniprot_1], 2)\n",
    "        evidence_df = relations_table[relations_table['agA_name'] == edge[0]]\n",
    "        evidence_df = evidence_df[evidence_df['agB_name'] == edge[1]]\n",
    "        evidence = evidence_df['evidence_count'].iloc[0]\n",
    "        if evidence >= EVIDENCE_FILTER and correlation > CORRELATION_FILTER:\n",
    "            params['color'] = '#00ff00'\n",
    "        else:\n",
    "            params['color'] = color\n",
    "        params['label'] = f'Correlation: {correlation}, Evidence: {evidence}'\n",
    "    ga.graph.add_edge(edge[0], edge[1], **params)\n",
    "\n",
    "ga.save_pdf(file_name='graph_annotations.pdf', prog='dot')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bda4aee-8c27-40a6-888a-9c822eb83728",
   "metadata": {},
   "source": [
    "### Example 2: Plotly Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cc63c3d1-aa86-4a9d-b80e-b74b40e263f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import random\n",
    "import networkx as nx\n",
    "import plotly.graph_objects as go\n",
    "from indra.databases import hgnc_client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0cc596a4-6ba0-4a96-b5d5-93c1f587e77b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def initialize_networkx_graph(subnetwork_relations, filter_bidirectional=False):\n",
    "    \"\"\"Return a networkx graph from the INDRA relations.\"\"\"\n",
    "    G = nx.DiGraph()\n",
    "    \n",
    "    # Get max evidence count of \"Complex\" INDRA statements for each pair of proteins.\n",
    "    ev_counts = {}\n",
    "    for entry in subnetwork_relations:\n",
    "        if entry['data']['stmt_type'] == \"Complex\":\n",
    "            if ev_counts.get((entry['source_id'], entry['target_id'])):\n",
    "                ev_counts[(entry['source_id'], entry['target_id'])] = \\\n",
    "                max(int(entry['data']['evidence_count']), int(ev_counts.get((entry['source_id'], entry['target_id']))))\n",
    "            else:\n",
    "                ev_counts[(entry['source_id'], entry['target_id'])] = int(entry['data']['evidence_count'])\n",
    "\n",
    "    # Construct graph object\n",
    "    for entry in subnetwork_relations:\n",
    "        source = entry['source_id']\n",
    "        target = entry['target_id']\n",
    "        \n",
    "        if filter_bidirectional:\n",
    "            # If there is a statement with opposite direction and more evidence\n",
    "            # then we skip this one\n",
    "            if ev_counts[(source, target)] < ev_counts.get((target, source), 0):\n",
    "                continue\n",
    "\n",
    "        # Add nodes to graph\n",
    "        source_name = hgnc_client.get_hgnc_name(source)\n",
    "        target_name = hgnc_client.get_hgnc_name(target)\n",
    "        G.add_node(source, label=source_name)\n",
    "        G.add_node(target, label=target_name)\n",
    "            \n",
    "        # Add the edge to the graph\n",
    "        G.add_edge(\n",
    "            source,\n",
    "            target,\n",
    "            evidence_count=ev_counts[(source, target)],\n",
    "            belief=entry['data']['belief'],\n",
    "            stmt_type=entry['data']['stmt_type']\n",
    "        )\n",
    "\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "38c43076-5acd-42ee-aec6-e6ce64b793d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_communities(G, weight='evidence_count'):\n",
    "    \"\"\"Return the communities of a networkx graph using a custom weight attribute.\"\"\"\n",
    "    return nx.community.louvain_communities(G, weight=weight)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f4f12aa7-b069-4fab-b08f-f62908330cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_node_initial_positions(G, communities):\n",
    "    \"\"\"Return node positions of a networkx graph based on communities.\"\"\"\n",
    "    initial_pos = {}\n",
    "    circle_r = 1\n",
    "    big_r = 1\n",
    "    pi = math.pi\n",
    "    centers = [(math.cos(2 * pi / len(communities) * x) * big_r, math.sin(2 * pi / len(communities) * x) * big_r)\n",
    "               for x in range(0, len(communities))]\n",
    "    for index, nodes in enumerate(communities):\n",
    "        for node in nodes:\n",
    "            alpha = 2 * math.pi * random.random()\n",
    "            r = circle_r * math.sqrt(random.random())\n",
    "            x = r * math.cos(alpha) + centers[index][0]\n",
    "            y = r * math.sin(alpha) + centers[index][1]\n",
    "            initial_pos[node] = [x, y]\n",
    "            nx.set_node_attributes(G, {node: index}, name='community')\n",
    "    \n",
    "    return initial_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eba35720-d0d5-4011-b6f9-a0bdd134c8cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_layout_to_graph(G, initial_pos, k=50, iterations=100, weight='evidence_count'):\n",
    "    \"\"\"Apply custom layout positions to a networkx graph.\"\"\"\n",
    "    pos = nx.spring_layout(G, weight=weight, k=k / math.sqrt(len(G.nodes)), pos=initial_pos, iterations=iterations)\n",
    "    for node in G.nodes():\n",
    "        x = pos[node][0]\n",
    "        y = pos[node][1]\n",
    "        nx.set_node_attributes(G, {node: [x, y]}, name='pos')\n",
    "    return G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "14662854-5f1b-486b-8696-71b0687abe08",
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_networkx_graph(subnetwork_relations):\n",
    "    \"\"\"Return a custom networkx graph from INDRA subnetwork relations.\"\"\"\n",
    "    G = initialize_networkx_graph(subnetwork_relations)\n",
    "    communities = find_communities(G)\n",
    "    initial_pos = generate_node_initial_positions(G, communities)\n",
    "    apply_layout_to_graph(G, initial_pos)\n",
    "    return G\n",
    "\n",
    "\n",
    "G = construct_networkx_graph(subnetwork_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ff27b5f8-de1a-4126-ad5c-12c7efef5669",
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_arrows(G):\n",
    "    \"\"\"Return list of directed edges for network visualization.\"\"\"\n",
    "    arrow_list = []\n",
    "    for edge in G.edges():\n",
    "        x0, y0 = G.nodes[edge[0]]['pos']\n",
    "        x1, y1 = G.nodes[edge[1]]['pos']\n",
    "\n",
    "        arrow = go.layout.Annotation(dict(\n",
    "            x=x0,\n",
    "            y=y0,\n",
    "            xref=\"x\", yref=\"y\",\n",
    "            showarrow=True,\n",
    "            axref=\"x\", ayref='y',\n",
    "            ax=x1,\n",
    "            ay=y1,\n",
    "            arrowhead=3,\n",
    "            arrowwidth=min(2, G.edges[edge]['evidence_count']),\n",
    "            arrowcolor='lightgreen')\n",
    "        )\n",
    "\n",
    "        arrow_list.append(arrow)\n",
    "    return arrow_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "da8e34d4-232a-4f08-afdb-df64de5c5b0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_arrow_labels(G):\n",
    "    \"\"\"Return custom edge labels for network visualization.\"\"\"\n",
    "    mnode_x, mnode_y, mnode_txt = [], [], []\n",
    "    for edge in G.edges():\n",
    "        x0, y0 = G.nodes[edge[0]]['pos']\n",
    "        x1, y1 = G.nodes[edge[1]]['pos']\n",
    "        name0 = G.nodes[edge[0]]['label']\n",
    "        name1 = G.nodes[edge[1]]['label']\n",
    "\n",
    "        mnode_x.extend([(x0 + x1) / 2])\n",
    "        mnode_y.extend([(y0 + y1) / 2])\n",
    "        mnode_txt.extend([f'{name0}->{name1} evidence count: {G.edges[edge][\"evidence_count\"]}'])\n",
    "\n",
    "    mnode_trace = go.Scatter(\n",
    "        x=mnode_x, y=mnode_y,\n",
    "        mode=\"markers\",\n",
    "        showlegend=False,\n",
    "        hovertext=mnode_txt,\n",
    "        hovertemplate=\"Edge %{hovertext}<extra></extra>\",\n",
    "        marker=go.scatter.Marker(opacity=0)\n",
    "    )\n",
    "    \n",
    "    return mnode_trace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7ef82dd0-da6b-4dd8-b33f-92338e31df3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def construct_node_trace(G):\n",
    "    \"\"\"Return custom nodes for network visualization\"\"\"\n",
    "    node_x = []\n",
    "    node_y = []\n",
    "    for node in G.nodes():\n",
    "        x, y = G.nodes[node]['pos']\n",
    "        node_x.append(x)\n",
    "        node_y.append(y)\n",
    "\n",
    "    node_trace = go.Scatter(\n",
    "        x=node_x, y=node_y,\n",
    "        mode='markers+text',\n",
    "        hoverinfo='text',\n",
    "        text=[data['label'] for node, data in list(G.nodes(data=True))],\n",
    "        textposition=\"bottom center\",\n",
    "        marker=dict(\n",
    "            showscale=True,\n",
    "            colorscale='YlGnBu',\n",
    "            reversescale=True,\n",
    "            color=[],\n",
    "            size=30,\n",
    "            colorbar=dict(\n",
    "                thickness=15,\n",
    "                title='Cluster ID',\n",
    "                xanchor='left',\n",
    "                titleside='right'\n",
    "            ),\n",
    "            line_width=2))\n",
    "    \n",
    "    node_colors = []\n",
    "    node_text = []\n",
    "    for node in G.nodes():\n",
    "        node_colors.append(G.nodes[node]['community'])\n",
    "\n",
    "    node_trace.marker.color = node_colors\n",
    "\n",
    "    return node_trace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8ccd9890-06ba-4121-b072-f7d810f8adc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_plotly_graph(nodes, arrows):\n",
    "    \"\"\"Visualize the network based on the list of nodes and arrows.\"\"\"\n",
    "    fig = go.Figure(data=nodes,\n",
    "                    layout=go.Layout(\n",
    "                        title='<br>Network graph made with Python',\n",
    "                        font=dict(\n",
    "                            family=\"Courier New, monospace\",\n",
    "                            size=10,\n",
    "                            color=\"Black\"\n",
    "                        ),\n",
    "                        annotations=arrows,\n",
    "                        showlegend=False,\n",
    "                        hovermode='closest',\n",
    "                        margin=dict(b=20, l=5, r=5, t=40),\n",
    "                        xaxis=dict(showgrid=False, zeroline=False, showticklabels=False),\n",
    "                        yaxis=dict(showgrid=False, zeroline=False, showticklabels=False))\n",
    "                    )\n",
    "    fig.show(renderer=\"notebook\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "af99ebc2-5785-4539-87c3-3abe56c82587",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_plotly_graph(G):\n",
    "    \"\"\"Generate and visualize a custom network from a networkx graph.\"\"\"\n",
    "    edges = construct_arrows(G)\n",
    "    edge_midpoints = construct_arrow_labels(G)\n",
    "    nodes = construct_node_trace(G)\n",
    "    show_plotly_graph([edge_midpoints, nodes], edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f640eede-a4ea-4df9-ade1-6048e11ddabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "create_plotly_graph(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c06ffaf8-007b-401f-8005-1d1b4c7acfa1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'label': 'BRD4',\n",
       " 'community': 0,\n",
       " 'pos': [0.9528133707816417, 0.19311777236587424]}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "G.nodes['13575']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
